<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    

    <title>Example 2: Differential expression scatterplots &mdash; metaseq 0.5.6 documentation</title>

<meta name="viewport" content="width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=0;"/>


    
    <link rel="stylesheet" href="_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/searchtools.js"></script>
    <link rel="top" title="metaseq 0.5.6 documentation" href="index.html" />
    <link rel="next" title="API docs" href="autodocs.html" />
    <link rel="prev" title="Example 1: Average ChIP-seq signal over promoters" href="example_session.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="autodocs.html" title="API docs"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="example_session.html" title="Example 1: Average ChIP-seq signal over promoters"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">metaseq 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="example-2-differential-expression-scatterplots">
<span id="example-session-2"></span><h1>Example 2: Differential expression scatterplots<a class="headerlink" href="#example-2-differential-expression-scatterplots" title="Permalink to this headline">¶</a></h1>
<p>This example looks more closely at using the results table part of
<code class="xref py py-mod docutils literal"><span class="pre">metaseq</span></code>, and highlights the flexibility in plotting afforded by
<code class="xref py py-mod docutils literal"><span class="pre">metaseq</span></code>.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Enable in-line plots for this IPython Notebook</span>
<span class="o">%</span><span class="n">matplotlib</span> <span class="n">inline</span>
</pre></div>
</div>
<div class="section" id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>In this section, we&#8217;ll get the example data for control and knockdown
samples, combine the data, and and create <code class="xref py py-class docutils literal"><span class="pre">ResultsTable</span></code> object
out of them.</p>
<p>If you haven&#8217;t already done so, run the
<cite>download_metaseq_example_data.py</cite> script, which will download and
prepare from public sources.</p>
<p>Import what we&#8217;ll be using:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">metaseq</span>
<span class="kn">from</span> <span class="nn">metaseq</span> <span class="k">import</span> <span class="n">example_filename</span>
<span class="kn">from</span> <span class="nn">metaseq.results_table</span> <span class="k">import</span> <span class="n">ResultsTable</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">gffutils</span>
<span class="kn">from</span> <span class="nn">gffutils.helpers</span> <span class="k">import</span> <span class="n">asinterval</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
<p>We&#8217;ll be using tables prepared from Cufflinks GTF output from GEO
entries GSM847565 and GSM847566. These represent results control and
ATF3 knockdown experiments in the K562 human cell line. You can read
more about the data on GEO; this example will be more about the features
of <code class="xref py py-mod docutils literal"><span class="pre">metaseq</span></code> than the biology.</p>
<p>Let&#8217;s get the example files:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span>%%bash
example_dir=&quot;metaseq-example&quot;

mkdir -p $example_dir
(cd $example_dir \
&amp;&amp; wget --progress=dot:giga https://raw.githubusercontent.com/daler/metaseq-example-data/master/metaseq-example-data.tar.gz \
&amp;&amp; tar -xzf metaseq-example-data.tar.gz \
&amp;&amp; rm metaseq-example-data.tar.gz)
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">--</span><span class="mi">2017</span><span class="o">-</span><span class="mi">01</span><span class="o">-</span><span class="mi">08</span> <span class="mi">10</span><span class="p">:</span><span class="mi">11</span><span class="p">:</span><span class="mi">20</span><span class="o">--</span>  <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">daler</span><span class="o">/</span><span class="n">metaseq</span><span class="o">-</span><span class="n">example</span><span class="o">-</span><span class="n">data</span><span class="o">/</span><span class="n">master</span><span class="o">/</span><span class="n">metaseq</span><span class="o">-</span><span class="n">example</span><span class="o">-</span><span class="n">data</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
<span class="n">Resolving</span> <span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span> <span class="p">(</span><span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span><span class="p">)</span><span class="o">...</span> <span class="mf">151.101</span><span class="o">.</span><span class="mf">32.133</span>
<span class="n">Connecting</span> <span class="n">to</span> <span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span> <span class="p">(</span><span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span><span class="p">)</span><span class="o">|</span><span class="mf">151.101</span><span class="o">.</span><span class="mf">32.133</span><span class="o">|</span><span class="p">:</span><span class="mf">443.</span><span class="o">..</span> <span class="n">connected</span><span class="o">.</span>
<span class="n">HTTP</span> <span class="n">request</span> <span class="n">sent</span><span class="p">,</span> <span class="n">awaiting</span> <span class="n">response</span><span class="o">...</span> <span class="mi">200</span> <span class="n">OK</span>
<span class="n">Length</span><span class="p">:</span> <span class="mi">96655384</span> <span class="p">(</span><span class="mi">92</span><span class="n">M</span><span class="p">)</span> <span class="p">[</span><span class="n">application</span><span class="o">/</span><span class="n">octet</span><span class="o">-</span><span class="n">stream</span><span class="p">]</span>
<span class="n">Saving</span> <span class="n">to</span><span class="p">:</span> <span class="s1">&#39;metaseq-example-data.tar.gz&#39;</span>

     <span class="mi">0</span><span class="n">K</span> <span class="o">........</span> <span class="o">........</span> <span class="o">........</span> <span class="o">........</span> <span class="mi">34</span><span class="o">%</span> <span class="mf">78.4</span><span class="n">M</span> <span class="mi">1</span><span class="n">s</span>
 <span class="mi">32768</span><span class="n">K</span> <span class="o">........</span> <span class="o">........</span> <span class="o">........</span> <span class="o">........</span> <span class="mi">69</span><span class="o">%</span>  <span class="mi">109</span><span class="n">M</span> <span class="mi">0</span><span class="n">s</span>
 <span class="mi">65536</span><span class="n">K</span> <span class="o">........</span> <span class="o">........</span> <span class="o">........</span> <span class="o">....</span>    <span class="mi">100</span><span class="o">%</span> <span class="mf">91.3</span><span class="n">M</span><span class="o">=</span><span class="mf">1.0</span><span class="n">s</span>

<span class="mi">2017</span><span class="o">-</span><span class="mi">01</span><span class="o">-</span><span class="mi">08</span> <span class="mi">10</span><span class="p">:</span><span class="mi">11</span><span class="p">:</span><span class="mi">21</span> <span class="p">(</span><span class="mf">91.2</span> <span class="n">MB</span><span class="o">/</span><span class="n">s</span><span class="p">)</span> <span class="o">-</span> <span class="s1">&#39;metaseq-example-data.tar.gz&#39;</span> <span class="n">saved</span> <span class="p">[</span><span class="mi">96655384</span><span class="o">/</span><span class="mi">96655384</span><span class="p">]</span>
</pre></div>
</div>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">data_dir</span> <span class="o">=</span> <span class="s1">&#39;metaseq-example/data&#39;</span>
<span class="n">control_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="s1">&#39;GSM847565_SL2585.table&#39;</span><span class="p">)</span>
<span class="n">knockdown_filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="s1">&#39;GSM847566_SL2592.table&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Let&#8217;s take a quick peak to see what these files look like:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span># System call; IPython only!
!head -n5 $control_filename
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="nb">id</span>  <span class="n">score</span>   <span class="n">fpkm</span>
<span class="n">ENST00000456328</span>     <span class="mf">108.293110992801</span>        <span class="mf">1.1183355144</span>
<span class="n">ENST00000515242</span>     <span class="mf">87.2330185289764</span>        <span class="mf">0.8306172562</span>
<span class="n">ENST00000518655</span>     <span class="mf">175.175608597549</span>        <span class="mf">2.3676823979</span>
<span class="n">ENST00000473358</span>     <span class="mf">343.232678519699</span>        <span class="mf">9.7952652359</span>
</pre></div>
</div>
<p>As documented at <a class="reference external" href="http://cufflinks.cbcb.umd.edu/manual.html#gtfout">http://cufflinks.cbcb.umd.edu/manual.html#gtfout</a>, the
<cite>score</cite> field indicates relative expression of one isoform compared to
other isoforms of the same gene, times 1000. The max score is 1000, and
an isoform with this score is considered the major isoform. A score of
800 would mean an isoform&#8217;s FPKM is 0.8 that of the major isoform.</p>
<p>If you&#8217;re working with DESeq results, the
<code class="xref py py-mod docutils literal"><span class="pre">metaseq.results_table.DESeqResults</span></code> class is a nice wrapper
around those results with one-step import. But here, we&#8217;ll construct a
<cite>pandas.DataFrame</cite> first and then create a <cite>ResultsTable</cite> object out
of it.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Create two pandas.DataFrames</span>

<span class="n">control</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">control_filename</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">knockdown</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">knockdown_filename</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Here&#8217;s what the first few entries look like:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">control</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>score</th>
      <th>fpkm</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENST00000456328</th>
      <td>108.293111</td>
      <td>1.118336</td>
    </tr>
    <tr>
      <th>ENST00000515242</th>
      <td>87.233019</td>
      <td>0.830617</td>
    </tr>
    <tr>
      <th>ENST00000518655</th>
      <td>175.175609</td>
      <td>2.367682</td>
    </tr>
    <tr>
      <th>ENST00000473358</th>
      <td>343.232679</td>
      <td>9.795265</td>
    </tr>
    <tr>
      <th>ENST00000408384</th>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div><div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">knockdown</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>score</th>
      <th>fpkm</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENST00000456328</th>
      <td>290.752543</td>
      <td>6.503301</td>
    </tr>
    <tr>
      <th>ENST00000515242</th>
      <td>253.364453</td>
      <td>4.790326</td>
    </tr>
    <tr>
      <th>ENST00000518655</th>
      <td>23.190962</td>
      <td>0.174388</td>
    </tr>
    <tr>
      <th>ENST00000473358</th>
      <td>510.475081</td>
      <td>33.409877</td>
    </tr>
    <tr>
      <th>ENST00000408384</th>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div><p>These are two separate objects. It will be easier to work with the data
if we first combine the data into a single dataframe. For this we will
use standard <cite>pandas</cite> routines:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Merge control and knockdown into one DataFrame</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">control</span><span class="p">,</span> <span class="n">knockdown</span><span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_ct&#39;</span><span class="p">,</span> <span class="s1">&#39;_kd&#39;</span><span class="p">))</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>score_ct</th>
      <th>fpkm_ct</th>
      <th>score_kd</th>
      <th>fpkm_kd</th>
    </tr>
    <tr>
      <th>id</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ENST00000456328</th>
      <td>108.293111</td>
      <td>1.118336</td>
      <td>290.752543</td>
      <td>6.503301</td>
    </tr>
    <tr>
      <th>ENST00000515242</th>
      <td>87.233019</td>
      <td>0.830617</td>
      <td>253.364453</td>
      <td>4.790326</td>
    </tr>
    <tr>
      <th>ENST00000518655</th>
      <td>175.175609</td>
      <td>2.367682</td>
      <td>23.190962</td>
      <td>0.174388</td>
    </tr>
    <tr>
      <th>ENST00000473358</th>
      <td>343.232679</td>
      <td>9.795265</td>
      <td>510.475081</td>
      <td>33.409877</td>
    </tr>
    <tr>
      <th>ENST00000408384</th>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div><p>Now we&#8217;ll create a <code class="xref py py-class docutils literal"><span class="pre">metaseq.results_table.ResultsTable</span></code> out of
it:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Create a ResultsTable</span>

<span class="n">d</span> <span class="o">=</span> <span class="n">ResultsTable</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<p><code class="xref py py-class docutils literal"><span class="pre">ResultsTable</span></code> objects are wrappers around <cite>pandas.DataFrame</cite>
objects, and are useful for working with annotations and tablular data.
You can always access the <cite>DataFrame</cite> with the <cite>.data</cite> attribute:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># DataFrame is always accessible via .data</span>

<span class="nb">print</span> <span class="nb">type</span><span class="p">(</span><span class="n">d</span><span class="p">),</span> <span class="nb">type</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>&lt;class &#39;metaseq.results_table.ResultsTable&#39;&gt; &lt;class &#39;pandas.core.frame.DataFrame&#39;&gt;
</pre></div>
</div>
<p>The <cite>metaseq</cite> example data includes a GFF file of the genes on
chromosome 17 of the hg19 human genome assembly:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Get gene annotations for chr17</span>

<span class="n">gtf</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="s1">&#39;Homo_sapiens.GRCh37.66_chr17.gtf&#39;</span><span class="p">)</span>
<span class="nb">print</span> <span class="nb">open</span><span class="p">(</span><span class="n">gtf</span><span class="p">)</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">chr17</span>       <span class="n">protein_coding</span>  <span class="n">exon</span>    <span class="mi">30898</span>   <span class="mi">31270</span>   <span class="o">.</span>       <span class="o">-</span>       <span class="o">.</span>        <span class="n">gene_id</span> <span class="s2">&quot;ENSG00000187939&quot;</span><span class="p">;</span> <span class="n">transcript_id</span> <span class="s2">&quot;ENST00000343572&quot;</span><span class="p">;</span> <span class="n">exon_number</span> <span class="s2">&quot;1&quot;</span><span class="p">;</span> <span class="n">gene_name</span> <span class="s2">&quot;DOC2B&quot;</span><span class="p">;</span> <span class="n">gene_biotype</span> <span class="s2">&quot;protein_coding&quot;</span><span class="p">;</span> <span class="n">transcript_name</span> <span class="s2">&quot;DOC2B-201&quot;</span><span class="p">;</span>
</pre></div>
</div>
</div>
<div class="section" id="subsetting-data">
<h2>Subsetting data<a class="headerlink" href="#subsetting-data" title="Permalink to this headline">¶</a></h2>
<p>The data we loaded from the knockdown experiment contains genes from all
chromosomes. For the sake of argument, let&#8217;s say we&#8217;re only interested
in the expression data for these genes on chr17. We can simply use
<cite>pandas.DataFrame.ix</cite> to subset dataframe by a list of genes. Note
that for this to work, the items in the list need to be in the index of
the dataframe. Since the data frame index consists of Ensembl transcript
IDs, we&#8217;ll need to create a list of Ensembl transcript IDs on chromosome
17:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Get a list of transcript IDs on chr17, and subset the dataframe.</span>

<span class="c1"># Here we use pybedtools, but the list of names can come from anywhere</span>
<span class="n">names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">i</span><span class="p">[</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">gtf</span><span class="p">)]))</span>
<span class="n">names</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

<span class="c1"># Make a copy of d</span>
<span class="n">d2</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

<span class="c1"># And subset</span>
<span class="n">d2</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">ix</span><span class="p">[</span><span class="n">names</span><span class="p">]</span>

<span class="c1"># How many did we omit?</span>
<span class="nb">print</span> <span class="s2">&quot;original:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
<span class="nb">print</span> <span class="s2">&quot;chr17 subset:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">original</span><span class="p">:</span> <span class="mi">85699</span>
<span class="n">chr17</span> <span class="n">subset</span><span class="p">:</span> <span class="mi">5708</span>
</pre></div>
</div>
</div>
<div class="section" id="scatterplots">
<h2>Scatterplots<a class="headerlink" href="#scatterplots" title="Permalink to this headline">¶</a></h2>
<p>Let&#8217;s plot some data. The <code class="xref py py-meth docutils literal"><span class="pre">ResultsTable.scatter()</span></code> method helps
with plotting genome-wide data, and offers lots of flexibility.</p>
<p>For its most basic usage, we need to at least supply <cite>x</cite> and <cite>y</cite>.
These are names of variables in the dataframe. We&#8217;ll add more data
later, but for now, let&#8217;s plot the FPKM of control vs knockdown:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Scatterplot of control vs knockdown FPKM</span>
<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_30_0.png" src="_images/example_session_2_30_0.png" />
<p>If you&#8217;re following along in a terminal with interactive <cite>matplotlib</cite>
plots, you can click on a point to see what gene it is. In this IPython
Notebook (and the HTML documentation generated from it), we don&#8217;t have
that interactive ability. We can simulate it here by choosing a gene ID
to show, and then manually call the <cite>_default_callback</cite> like this:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># arbitrary gene for demonstration purposes</span>
<span class="n">interesting_gene</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">fpkm_ct</span><span class="p">)</span>
<span class="n">interesting_gene</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="s1">&#39;ENST00000253788&#39;</span>
</pre></div>
</div>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># What happens if you were to click on the points in an interactive session</span>
<span class="n">d2</span><span class="o">.</span><span class="n">_default_callback</span><span class="p">(</span><span class="n">interesting_gene</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">score_ct</span>    <span class="mf">1047.517457</span>
<span class="n">fpkm_ct</span>     <span class="mf">1422.448488</span>
<span class="n">score_kd</span>    <span class="mf">1070.752317</span>
<span class="n">fpkm_kd</span>     <span class="mf">1671.190119</span>
<span class="n">Name</span><span class="p">:</span> <span class="n">ENST00000253788</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">float64</span>
</pre></div>
</div>
<p>Clicking around interactively on the points is a great way to get a feel
for the data.</p>
<p>OK, it looks like this plot could use log scaling. Recall though that
the <cite>ResultsTable.scatter</cite> method needs to have <cite>x</cite> and <cite>y</cite>
variables available in the dataframe. So one way to do this would be to
do something like this:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Adding extra variables gets verbose and cluttered</span>
<span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;log_fpkm_ct&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">fpkm_ct</span><span class="p">)</span>
</pre></div>
</div>
<p>But when playing around with different scales, this quickly pollutes the
dataframe with extra columns. Let&#8217;s delete that column . . .</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># We&#39;ll use a better way, so delete it.</span>
<span class="k">del</span> <span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;log_fpkm_ct&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>. . . and show another way.</p>
<p>You may find it more streamlined to use the <cite>xfunc</cite> and/or <cite>yfunc</cite>
arguments. We can use any arbitrary function for these, and the axes
labels will reflect that.</p>
<p>Since we&#8217;re about to start incrementally improving the figure by adding
additional keyword arguments (kwargs), the stuff we&#8217;ve already talked
about will be at the top, and a comment line like this will mark the
start of new stuff to pay attention to:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># -------------  (marks the start of new stuff)</span>
</pre></div>
</div>
<p>Here&#8217;s the next version of the scatterplot:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Scale x and y axes using log2(x + 1)</span>

<span class="k">def</span> <span class="nf">log2p1</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="c1">#----------------</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_39_0.png" src="_images/example_session_2_39_0.png" />
<p>Of course, we can specify axes labels either directly in the method call
with <cite>xlab</cite> or <cite>ylab</cite>, or after the fact using standard
<cite>matplotlib</cite> functionality:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Manually specify x and y labels</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>

    <span class="c1">#-----------------------------</span>
    <span class="c1"># specify xlabel</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span>
    <span class="p">);</span>

<span class="c1"># adjust the ylabel afterwards</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_41_0.png" src="_images/example_session_2_41_0.png" />
<p>Let&#8217;s highlight some genes. How about those that change expression &gt; 2
fold in upon knockdown in red, and &lt; 2 fold in blue? While we&#8217;re at it,
let&#8217;s add another variable to the dataframe.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Crude differential expression detection....</span>

<span class="n">d2</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;foldchange&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">fpkm_kd</span> <span class="o">/</span> <span class="n">d2</span><span class="o">.</span><span class="n">fpkm_ct</span>
<span class="n">up</span> <span class="o">=</span> <span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">foldchange</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="n">dn</span> <span class="o">=</span> <span class="p">(</span><span class="n">d2</span><span class="o">.</span><span class="n">foldchange</span> <span class="o">&lt;</span> <span class="mf">0.5</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
<p>The way to highlight genes is with the <cite>genes_to_highlight</cite> argument.
OK, OK, it&#8217;s a little bit of a misnomer here because we&#8217;re actually
working with transcripts. But the idea is the same.</p>
<p>The <cite>genes_to_highlight</cite> argument takes a list of tuples. Each tuple
consists of two items: an index (boolean or integer, doesn&#8217;t matter) and
a style dictionary. This dictionary is passed directly to
<cite>matplotlib.scatter</cite>, so you can use any supported arguments here.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">There are actually other kwargs you can use in the style dictionary &#8211; for example, <cite>names</cite>, <cite>marginal_histograms</cite>, <cite>xhist_kwargs</cite>, and <cite>yhist_kwargs</cite>.  These are <cite>metaseq</cite>-specific and will be explained later.</p>
</div>
<p>Here&#8217;s the plot with up/downregulated genes highlighted:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Use the genes_to_highlight argument to show up/downregulated genes</span>
<span class="c1"># in different colors</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="c1">#-------------------------------</span>
    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="n">up</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">)),</span>
        <span class="p">(</span><span class="n">dn</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">))]</span>
    <span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_47_0.png" src="_images/example_session_2_47_0.png" />
<p>We can add a 1-to-1 line for reference:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Add a 1:1 line</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="n">up</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">)),</span>
        <span class="p">(</span><span class="n">dn</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">))],</span>
    <span class="c1">#------------------------------------------</span>
    <span class="n">one_to_one</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">),</span>
    <span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_49_0.png" src="_images/example_session_2_49_0.png" />
<p>Let&#8217;s change the plot style a bit. The <cite>general_kwargs</cite> argument
determines the base style of all points. By default, it&#8217;s
<cite>dict(color=&#8217;k&#8217;, alpha=0.2, linewidths=0)</cite>. Let&#8217;s change the default
style to smaller gray dots, and make the red and blue stand out more by
adjusting their alpha:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Style changes:</span>
<span class="c1"># default gray small dots; make changed genes stand out more</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">one_to_one</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">),</span>
    <span class="c1">#------------------------------------------------------</span>
    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="n">up</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)),</span>
        <span class="p">(</span><span class="n">dn</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">))],</span>
    <span class="n">general_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;0.5&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">),</span>
    <span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_51_0.png" src="_images/example_session_2_51_0.png" />
</div>
<div class="section" id="marginal-histograms">
<h2>Marginal histograms<a class="headerlink" href="#marginal-histograms" title="Permalink to this headline">¶</a></h2>
<p><code class="xref py py-mod docutils literal"><span class="pre">metaseq</span></code> also offers support for marginal histograms, which are
stacked up on either axes for each set of genes that were plotted. There
are lots of ways for configuring this. First, let&#8217;s turn them on for
everything:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Add marginal histograms</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="n">up</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)),</span>
        <span class="p">(</span><span class="n">dn</span><span class="p">,</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">))],</span>
    <span class="n">one_to_one</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">),</span>
    <span class="n">general_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;0.5&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">),</span>
    <span class="c1">#------------------------------------------------------</span>
    <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_53_0.png" src="_images/example_session_2_53_0.png" />
<p>As a contrived example to illustrate the flexibility for plotting
marginal histograms, lets:</p>
<ul class="simple">
<li>only show histograms for up/down regulated</li>
<li>change the number of bins to 50</li>
<li>remove the edge around each bar</li>
</ul>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Tweak the marginal histograms:</span>
<span class="c1"># 50 bins, don&#39;t show unchanged genes, and remove outlines</span>

<span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">one_to_one</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">),</span>
    <span class="n">general_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;0.5&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">),</span>
    <span class="c1">#------------------------------------------------------</span>
    <span class="c1"># Go back go disabling them globally...</span>
    <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>

    <span class="c1"># ...and then turn them back on for each set of genes</span>
    <span class="c1"># to highlight.</span>
    <span class="c1">#</span>
    <span class="c1"># By the way, genes_to_highlight is indented to better show the</span>
    <span class="c1"># the structure.</span>
    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
    <span class="p">(</span>
     <span class="n">up</span><span class="p">,</span>
     <span class="nb">dict</span><span class="p">(</span>
          <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
          <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
          <span class="n">xhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="n">yhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="p">)</span>
     <span class="p">),</span>
    <span class="p">(</span>
     <span class="n">dn</span><span class="p">,</span>
     <span class="nb">dict</span><span class="p">(</span>
          <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
          <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
          <span class="n">xhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="n">yhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="p">)</span>
     <span class="p">)</span>
    <span class="p">],</span>
<span class="p">);</span>
</pre></div>
</div>
<img alt="_images/example_session_2_55_0.png" src="_images/example_session_2_55_0.png" />
<p>Let&#8217;s clean up the plot by adding a legend (using <cite>label</cite> in
<cite>genes_to_highlight</cite>), and adding it outside the axes. While we&#8217;re at
it we&#8217;ll add a title, too.</p>
<p>There&#8217;s a trick here &#8211; for each set of genes, the histograms are
incrementally added on top of each other but the legend, lists them
going down. So we need to flip the order of legend entries to make it
nicely match the order of the histograms.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;font.family&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;Arial&quot;</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="s1">&#39;fpkm_ct&#39;</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="s1">&#39;fpkm_kd&#39;</span><span class="p">,</span>
    <span class="n">xfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">yfunc</span><span class="o">=</span><span class="n">log2p1</span><span class="p">,</span>
    <span class="n">xlab</span><span class="o">=</span><span class="s1">&#39;Control, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">ylab</span><span class="o">=</span><span class="s1">&#39;Knockdown, log2(FPKM + 1)&#39;</span><span class="p">,</span>
    <span class="n">one_to_one</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">),</span>
    <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="c1">#------------------------------------------------------</span>
    <span class="c1"># add the &quot;unchanged&quot; label</span>
    <span class="n">general_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;0.5&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;unchanged&#39;</span><span class="p">),</span>

    <span class="n">genes_to_highlight</span><span class="o">=</span><span class="p">[</span>
    <span class="p">(</span>
     <span class="n">up</span><span class="p">,</span>
     <span class="nb">dict</span><span class="p">(</span>
          <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#da3b3a&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
          <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
          <span class="n">xhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="n">yhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>

          <span class="c1"># add label</span>
          <span class="n">label</span><span class="o">=</span><span class="s1">&#39;upregulated&#39;</span><span class="p">,</span>
          <span class="p">)</span>
     <span class="p">),</span>
    <span class="p">(</span>
     <span class="n">dn</span><span class="p">,</span>
     <span class="nb">dict</span><span class="p">(</span>
          <span class="n">color</span><span class="o">=</span><span class="s1">&#39;#00748e&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
          <span class="n">marginal_histograms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
          <span class="n">xhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
          <span class="n">yhist_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>

          <span class="c1"># add label</span>
          <span class="n">label</span><span class="o">=</span><span class="s1">&#39;downregulated&#39;</span>
          <span class="p">)</span>
     <span class="p">)</span>
    <span class="p">],</span>
<span class="p">);</span>

<span class="c1"># Get handles and labels, and then reverse their order</span>
<span class="n">handles</span><span class="p">,</span> <span class="n">legend_labels</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_legend_handles_labels</span><span class="p">()</span>
<span class="n">handles</span> <span class="o">=</span> <span class="n">handles</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">legend_labels</span> <span class="o">=</span> <span class="n">legend_labels</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># Draw a legend using the flipped handles and labels.</span>
<span class="n">leg</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="p">,</span>
          <span class="n">legend_labels</span><span class="p">,</span>

          <span class="c1"># These values may take some tweaking.</span>
          <span class="c1"># By default they are in axes coordinates, so this means</span>
          <span class="c1"># the legend is slightly outside the axes.</span>
          <span class="n">loc</span><span class="o">=</span><span class="p">(</span><span class="mf">1.01</span><span class="p">,</span> <span class="mf">1.05</span><span class="p">),</span>

          <span class="c1"># Various style fixes to default legend.</span>
          <span class="n">fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span>
          <span class="n">scatterpoints</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
          <span class="n">borderpad</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
          <span class="n">handletextpad</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
          <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
          <span class="n">title</span><span class="o">=</span><span class="s1">&#39;chr17 transcripts&#39;</span><span class="p">,</span>
          <span class="p">);</span>

<span class="c1"># Adjust the legend title after it&#39;s created</span>
<span class="n">leg</span><span class="o">.</span><span class="n">get_title</span><span class="p">()</span><span class="o">.</span><span class="n">set_weight</span><span class="p">(</span><span class="s1">&#39;bold&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img alt="_images/example_session_2_57_0.png" src="_images/example_session_2_57_0.png" />
<p>We&#8217;d also like to add a title. But how to access the top-most axes?</p>
<p>Whenever the <cite>scatter</cite> method is called, the <cite>MarginalHistograms</cite>
object created as a by-product of the plotting is stored in the
<cite>marginal</cite> attribute. This, in turn, has a <cite>top_hists</cite> attribute,
and we can grab the last one created. While we&#8217;re at it, let&#8217;s
histograms axes as well.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1"># When `d2.scatter` is called, we get a `marginal` attribute.</span>
<span class="n">top_axes</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">marginal</span><span class="o">.</span><span class="n">top_hists</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
<span class="n">top_axes</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Differential expression, ATF3 knockdown&#39;</span><span class="p">);</span>

<span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">d2</span><span class="o">.</span><span class="n">marginal</span><span class="o">.</span><span class="n">top_hists</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;No.</span><span class="se">\n</span><span class="s1">transcripts&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="n">d2</span><span class="o">.</span><span class="n">marginal</span><span class="o">.</span><span class="n">right_hists</span><span class="p">:</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;No.</span><span class="se">\n</span><span class="s1">transcripts&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=-</span><span class="mi">90</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">figure</span>
<span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;expression-demo.png&#39;</span><span class="p">)</span>
<span class="n">fig</span>
</pre></div>
</div>
<img alt="_images/example_session_2_59_0.png" src="_images/example_session_2_59_0.png" />
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Example 2: Differential expression scatterplots</a><ul>
<li><a class="reference internal" href="#setup">Setup</a></li>
<li><a class="reference internal" href="#subsetting-data">Subsetting data</a></li>
<li><a class="reference internal" href="#scatterplots">Scatterplots</a></li>
<li><a class="reference internal" href="#marginal-histograms">Marginal histograms</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="example_session.html"
                        title="previous chapter">Example 1: Average ChIP-seq signal over promoters</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="autodocs.html"
                        title="next chapter">API docs</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/example_session_2.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="autodocs.html" title="API docs"
             >next</a> |</li>
        <li class="right" >
          <a href="example_session.html" title="Example 1: Average ChIP-seq signal over promoters"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">metaseq 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2015, Ryan Dale.
  Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.4.1.
  <br />Theme based on <a href="http://readthedocs.org/">Read The Docs</a>

</div>





  </body>
</html>