<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    

    <title>Metaseq &mdash; metaseq 0.5.6 documentation</title>

<meta name="viewport" content="width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=0;"/>


    
    <link rel="stylesheet" href="_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.5.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/searchtools.js"></script>
    <link rel="top" title="metaseq 0.5.6 documentation" href="#" />
    <link rel="next" title="Installation" href="install.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="install.html" title="Installation"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">metaseq 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="metaseq">
<h1>Metaseq<a class="headerlink" href="#metaseq" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://travis-ci.org/daler/metaseq"><img alt="https://travis-ci.org/daler/metaseq.png?branch=master" src="https://travis-ci.org/daler/metaseq.png?branch=master" /></a>
<a class="reference external image-reference" href="http://badge.fury.io/py/metaseq"><img src="https://badge.fury.io/py/metaseq.svg" /></a>
<a class="reference external image-reference" href="http://bioconda.github.io"><img alt="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square" src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square" /></a>
<p>Briefly, the goal of <cite>metaseq</cite> is to tie together lots of existing software into
a framework for exploring genomic data.  It focuses on flexibility and
interactive exploration and plotting of disparate genomic data sets.</p>
<p>The main documentation for <cite>metaseq</cite> can be found at <a class="reference external" href="https://daler.github.io/metaseq">https://daler.github.io/metaseq</a>.</p>
<p>If you use <cite>metaseq</cite> in your work, please cite the following publication:</p>
<blockquote>
<div>Dale, R. K., Matzat, L. H. &amp; Lei, E. P. metaseq: a Python package for
integrative genome-wide analysis reveals relationships between chromatin
insulators and associated nuclear mRNA. Nucleic Acids Res. 42, 9158–9170
(2014). <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/25063299">http://www.ncbi.nlm.nih.gov/pubmed/25063299</a></div></blockquote>
<div class="section" id="example-1-average-chip-seq-signal-over-promoters">
<h2>Example 1: Average ChIP-seq signal over promoters<a class="headerlink" href="#example-1-average-chip-seq-signal-over-promoters" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="example_session_run.html#example-session"><span class="std std-ref">Example 1</span></a> walks you
through the creation of the following heatmap and line-plot figure:</p>
<div class="figure" id="id1">
<img alt="_images/demo.png" src="_images/demo.png" />
<p class="caption"><span class="caption-text">Top: Heatmap of ATF3 ChIP-seq signal over transcription start sites (TSS) on
chr17 in human K562 cells.  Middle: average ChIP enrichment over all TSSs
+/- 1kb, with 95% CI band.  Bottom: Integration with ATF3 knockdown RNA-seq
results, showing differential enrichment over transcripts that went up,
down, or were unchanged upon ATF3 knockdown.</span></p>
</div>
</div>
<div class="section" id="example-2-differential-expression-scatterplots">
<h2>Example 2: Differential expression scatterplots<a class="headerlink" href="#example-2-differential-expression-scatterplots" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="example_session_2.html#example-session-2"><span class="std std-ref">Example 2</span></a> walks
you through the creation of the following scatterplot and marginal histogram
figure:</p>
<div class="figure" id="id2">
<img alt="_images/expression-demo.png" src="_images/expression-demo.png" />
<p class="caption"><span class="caption-text">Control vs knockdown expression (log2(FPKM + 1)) for an ATF3 knockdown
experiment.  Each point represents one transcript on chromosome 17.
Marginal distributions are shown on top and side.  1:1 line shown as
a dotted line.  Up- and downregulated genes determined by a simple 2-fold
cutoff.</span></p>
</div>
</div>
<div class="section" id="other-features">
<h2>Other features<a class="headerlink" href="#other-features" title="Permalink to this headline">¶</a></h2>
<p>In addition, <cite>metaseq</cite> offers:</p>
<ul class="simple">
<li>A format-agnostic API for accessing &#8220;genomic signal&#8221; that allows you to work
with BAM, BED, VCF, GTF, GFF, bigBed, and bigWig using the same API.</li>
<li>Parallel data access from the file formats mentioned above</li>
<li>&#8220;Mini-browsers&#8221;, zoomable and pannable Python-only  figures that show genomic
signal and gene models and are spawned by clicking on features of interest</li>
<li>A wrapper around pandas.DataFrames to simplify the manipulation and plotting
of tabular results data that contain gene information (like DESeq results
tables)</li>
<li>Integrates data keyed by genomic interval (think BAM or BED files) with data
keyed by gene ID (e.g., Cufflinks or DESeq results tables)</li>
</ul>
<p>Check out the <a class="reference external" href="https://daler.github.io/metaseq/">full documentation</a> for
more.</p>
<p>Some use-cases for <code class="xref py py-mod docutils literal"><span class="pre">metaseq</span></code>:</p>
<ul class="simple">
<li>Create NumPy arrays (which can be plotted as heatmaps) where each row
represents a feature (say, TSS +/- 1kb) at fairly high speed.  The actual
speed is highly dependent on your data, but a rule of thumb is that it
takes about 10s for every 10,000 2-kb features.  Once you have the heatmap,
you can sort, cluster, zoom, and pan to explore interesting groups.  See
the <span class="xref std std-ref">CTCF example</span> for more on this as well as the
<code class="xref py py-mod docutils literal"><span class="pre">metaseq.genomic_signal</span></code> module.</li>
<li>&#8220;Mini-browsers&#8221;.  Instead of tweaking something, then laboriously converting
to bedGraph, then bigWig, uploading your tracks to the genome browser, then
searching for the region of interest, you can can just connect a mini-browser
to your raw data and spawn figures locally.  See the <a class="reference internal" href="example_session.html#example-session"><span class="std std-ref">Example 1: Average ChIP-seq signal over promoters</span></a> for
more on this as well as the <code class="xref py py-mod docutils literal"><span class="pre">metaseq.minibrowser</span></code> API docs.</li>
<li>Cluster genes based on the spatial distribution of ChIP-seq peaks around
their TSSs.</li>
<li>Scatter plot of DESeq results (basemeana vs basemeanb) where points are
colored according to the number of ChIP peaks in the gene.  This, too, can be
attached to mini-browsers, enabling you to click on a point to see the
genomic signal.  See the <code class="xref py py-mod docutils literal"><span class="pre">metaseq.results_table</span></code> module for more on
this.</li>
<li>Pie charts of where peaks fall within annotated genes &#8211; TSS, poly-A
site, intron, exon, etc.  See the <code class="xref py py-mod docutils literal"><span class="pre">metaseq.integration</span></code> module for more
on this.</li>
</ul>
<p>Where possible, the inputs are standard formats &#8211; BED, GFF, GTF, BAM, SAM,
DESeq results as saved from R, or even arbitrary tab-delimited data files that
have a header.  If you take the time to convert to bigWig or bigBed,
performance will be improved.</p>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="install.html#easy-installation-method">Easy installation method</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="running-the-examples.html">Running the examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_session.html">Example 1: Average ChIP-seq signal over promoters</a><ul>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#example-data">Example data</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#getting-tsses">Getting TSSes</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#creating-the-arrays">Creating the arrays</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#loading-the-arrays">Loading the arrays</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#line-plot-of-average-signal">Line plot of average signal</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#adding-a-heatmap">Adding a heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#sorting-the-array">Sorting the array</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#customizing-the-axes-styles">Customizing the axes styles</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_session.html#integrating-with-rna-seq-expression-data">Integrating with RNA-seq expression data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#rna-seq-data-wrangling-loading-data">RNA-seq data wrangling: loading data</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#rna-seq-data-wrangling-aligning-rna-seq-data-with-chip-seq-data">RNA-seq data wrangling: aligning RNA-seq data with ChIP-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#rna-seq-data-wrangling-join-control-and-knockdown-data">RNA-seq data wrangling: join control and knockdown data</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#integrating-rna-seq-data-with-the-heatmap">Integrating RNA-seq data with the heatmap</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_session.html#extras">Extras</a><ul>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#k-means-clustering-of-chip-seq-signal">K-means clustering of ChIP-seq signal</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session.html#scatterplots-of-rna-seq-and-chip-seq-signal">Scatterplots of RNA-seq and ChIP-seq signal</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="example_session_2.html">Example 2: Differential expression scatterplots</a><ul>
<li class="toctree-l2"><a class="reference internal" href="example_session_2.html#setup">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session_2.html#subsetting-data">Subsetting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session_2.html#scatterplots">Scatterplots</a></li>
<li class="toctree-l2"><a class="reference internal" href="example_session_2.html#marginal-histograms">Marginal histograms</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="autodocs.html">API docs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#genomic-signal"><code class="docutils literal"><span class="pre">genomic_signal</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-results-table"><code class="docutils literal"><span class="pre">metaseq.results_table</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-integration"><code class="docutils literal"><span class="pre">metaseq.integration</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-plotutils"><code class="docutils literal"><span class="pre">metaseq.plotutils</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-colormap-adjust"><code class="docutils literal"><span class="pre">metaseq.colormap_adjust</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-minibrowser"><code class="docutils literal"><span class="pre">metaseq.minibrowser</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="autodocs.html#metaseq-filetype-adapters"><code class="docutils literal"><span class="pre">metaseq.filetype_adapters</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changes.html">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="changes.html#changes-in-v0-5-6">Changes in v0.5.6</a></li>
<li class="toctree-l2"><a class="reference internal" href="changes.html#changes-in-v0-5-5-1">Changes in v0.5.5.1</a></li>
<li class="toctree-l2"><a class="reference internal" href="changes.html#changes-in-v0-5-5">Changes in v0.5.5</a></li>
<li class="toctree-l2"><a class="reference internal" href="changes.html#changes-in-v0-5-4">Changes in v0.5.4</a></li>
<li class="toctree-l2"><a class="reference internal" href="changes.html#changes-in-v0-5">Changes in v0.5</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Metaseq</a><ul>
<li><a class="reference internal" href="#example-1-average-chip-seq-signal-over-promoters">Example 1: Average ChIP-seq signal over promoters</a></li>
<li><a class="reference internal" href="#example-2-differential-expression-scatterplots">Example 2: Differential expression scatterplots</a></li>
<li><a class="reference internal" href="#other-features">Other features</a></li>
</ul>
</li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="install.html"
                        title="next chapter">Installation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="install.html" title="Installation"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">metaseq 0.5.6 documentation</a> &raquo;</li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2015, Ryan Dale.
  Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.4.1.
  <br />Theme based on <a href="http://readthedocs.org/">Read The Docs</a>

</div>





  </body>
</html>