<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    

    <title>metaseq.integration.chipseq &mdash; metaseq 0.5.6 documentation</title>

<meta name="viewport" content="width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=0;"/>


    
    <link rel="stylesheet" href="../../../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.5.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../../../../_static/searchtools.js"></script>
    <link rel="top" title="metaseq 0.5.6 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">metaseq 0.5.6 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for metaseq.integration.chipseq</h1><div class="highlight"><pre>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module integrates parts of metaseq that are useful for ChIP-seq analysis.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">izip</span>
<span class="kn">import</span> <span class="nn">gffutils</span>
<span class="kn">from</span> <span class="nn">gffutils.helpers</span> <span class="kn">import</span> <span class="n">asinterval</span>
<span class="kn">import</span> <span class="nn">metaseq</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">from</span> <span class="nn">metaseq.minibrowser</span> <span class="kn">import</span> <span class="n">SignalMiniBrowser</span><span class="p">,</span> <span class="n">GeneModelMiniBrowser</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">yaml</span>

<span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">prefix</span><span class="p">,</span> <span class="n">relative_paths</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Save data from a Chipseq object.</span>


<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    c : Chipseq object</span>
<span class="sd">        Chipseq object, most likely after calling the `diffed_array` method</span>

<span class="sd">    prefix : str</span>
<span class="sd">        Prefix, including any leading directory paths, to save the data.</span>

<span class="sd">    relative_paths : bool</span>
<span class="sd">        If True (default), then the path names in the `prefix.info` file will</span>
<span class="sd">        be relative to `prefix`.  Otherwise, they will be absolute.</span>

<span class="sd">    The following files will be created:</span>

<span class="sd">    :prefix.intervals:</span>
<span class="sd">        A BED file (or GFF, GTF, or VCF as appropriate) of the features used for the array</span>

<span class="sd">    :prefix.info:</span>
<span class="sd">        A YAML-format file indicating the IP and control BAM files, any array</span>
<span class="sd">        kwargs, the database filename, and any minibrowser local coverage args.</span>
<span class="sd">        These are all needed to reconstruct a new Chipseq object. Path names</span>
<span class="sd">        will be relative to `prefix`.</span>

<span class="sd">    :prefix.npz:</span>
<span class="sd">        A NumPy .npz file with keys &#39;diffed_array&#39;, &#39;ip_array&#39;, and &#39;control_array&#39;</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">dirname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span>

    <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">features</span><span class="p">)</span><span class="o">.</span><span class="n">saveas</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&#39;.intervals&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">usepath</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">relative_paths</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">relpath</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">dirname</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&#39;.info&#39;</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fout</span><span class="p">:</span>
        <span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s">&#39;ip_bam&#39;</span><span class="p">:</span> <span class="n">usepath</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">ip</span><span class="o">.</span><span class="n">fn</span><span class="p">),</span>
            <span class="s">&#39;control_bam&#39;</span><span class="p">:</span> <span class="n">usepath</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">control</span><span class="o">.</span><span class="n">fn</span><span class="p">),</span>
            <span class="s">&#39;array_kwargs&#39;</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">array_kwargs</span><span class="p">,</span>
            <span class="s">&#39;dbfn&#39;</span><span class="p">:</span> <span class="n">usepath</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">dbfn</span><span class="p">),</span>
            <span class="s">&#39;browser_local_coverage_kwargs&#39;</span><span class="p">:</span> <span class="n">c</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span><span class="p">,</span>
            <span class="s">&#39;relative_paths&#39;</span><span class="p">:</span> <span class="n">relative_paths</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">fout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">yaml</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">info</span><span class="p">,</span> <span class="n">default_flow_style</span><span class="o">=</span><span class="bp">False</span><span class="p">))</span>
    <span class="n">np</span><span class="o">.</span><span class="n">savez</span><span class="p">(</span>
        <span class="n">prefix</span><span class="p">,</span>
        <span class="n">diffed_array</span><span class="o">=</span><span class="n">c</span><span class="o">.</span><span class="n">diffed_array</span><span class="p">,</span>
        <span class="n">ip_array</span><span class="o">=</span><span class="n">c</span><span class="o">.</span><span class="n">ip_array</span><span class="p">,</span>
        <span class="n">control_array</span><span class="o">=</span><span class="n">c</span><span class="o">.</span><span class="n">control_array</span>
    <span class="p">)</span>

<span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="n">prefix</span><span class="p">):</span>
    <span class="n">info</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&#39;.info&#39;</span><span class="p">))</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">Chipseq</span><span class="p">(</span><span class="n">ip_bam</span><span class="o">=</span><span class="n">info</span><span class="p">[</span><span class="s">&#39;ip_bam&#39;</span><span class="p">],</span> <span class="n">control_bam</span><span class="o">=</span><span class="n">info</span><span class="p">[</span><span class="s">&#39;control_bam&#39;</span><span class="p">],</span>
                <span class="n">dbfn</span><span class="o">=</span><span class="n">info</span><span class="p">[</span><span class="s">&#39;dbfn&#39;</span><span class="p">])</span>
    <span class="n">npz</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&#39;.npz&#39;</span><span class="p">,</span> <span class="n">mmap_mode</span><span class="o">=</span><span class="s">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">c</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">=</span> <span class="n">npz</span><span class="p">[</span><span class="s">&#39;ip_array&#39;</span><span class="p">]</span>
    <span class="n">c</span><span class="o">.</span><span class="n">control_array</span> <span class="o">=</span> <span class="n">npz</span><span class="p">[</span><span class="s">&#39;control_array&#39;</span><span class="p">]</span>
    <span class="n">c</span><span class="o">.</span><span class="n">diffed_array</span> <span class="o">=</span> <span class="n">npz</span><span class="p">[</span><span class="s">&#39;diffed_array&#39;</span><span class="p">]</span>
    <span class="n">c</span><span class="o">.</span><span class="n">array_kwargs</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="s">&#39;array_kwargs&#39;</span><span class="p">]</span>
    <span class="n">c</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="s">&#39;.intervals&#39;</span><span class="p">)</span>
    <span class="n">c</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span> <span class="o">=</span> <span class="n">info</span><span class="p">[</span><span class="s">&#39;browser_local_coverage_kwargs&#39;</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">c</span>

<span class="k">class</span> <span class="nc">GeneChipseqMiniBrowser</span><span class="p">(</span><span class="n">GeneModelMiniBrowser</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="n">db</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">GeneChipseqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span>
            <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="n">db</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">GeneChipseqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">axes</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s">&#39;best&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;RPMMR&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_major_formatter</span><span class="p">(</span>
            <span class="n">matplotlib</span><span class="o">.</span><span class="n">ticker</span><span class="o">.</span><span class="n">FormatStrFormatter</span><span class="p">(</span><span class="s">&#39;</span><span class="si">%d</span><span class="s">&#39;</span><span class="p">))</span>
        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span>


<span class="k">class</span> <span class="nc">SignalChipseqMiniBrowser</span><span class="p">(</span><span class="n">SignalMiniBrowser</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">SignalChipseqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span>
            <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">SignalChipseqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">ax1</span><span class="p">,</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">axes</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s">&#39;best&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">&#39;RPMMR&#39;</span><span class="p">)</span>
        <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span>


<div class="viewcode-block" id="Chipseq"><a class="viewcode-back" href="../../../autodocs/metaseq.integration.chipseq.Chipseq.html#metaseq.integration.chipseq.Chipseq">[docs]</a><span class="k">class</span> <span class="nc">Chipseq</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class for visualizing and interactively exploring ChIP-seq data.</span>

<span class="sd">    Needs two BAM files (for IP and control) and a gffutils database</span>
<span class="sd">    filename in order to display gene models.</span>

<span class="sd">    Typical usage is to create a normalized array of signal over each</span>
<span class="sd">    feature with the `diff_array` method, and then plot with the `plot`</span>
<span class="sd">    method.</span>

<span class="sd">    The resulting figure has the matrix as a heatmap, the average signal</span>
<span class="sd">    over features, and a panel with points that can be zoomed and clicked,</span>
<span class="sd">    spawning a mini-browser window for the corresponding feature.</span>

<span class="sd">    Configuration can be done by adjusting the following attributes after</span>
<span class="sd">    creating a Chipseq instance:</span>

<span class="sd">        * `_strip_kwargs` (the style for the dots in the left panel)</span>
<span class="sd">        * `browser_plotting_kwargs` (style of signal lines in the mini-browser)</span>

<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="Chipseq.__init__"><a class="viewcode-back" href="../../../autodocs/metaseq.integration.chipseq.Chipseq.html#metaseq.integration.chipseq.Chipseq.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ip_bam</span><span class="p">,</span> <span class="n">control_bam</span><span class="p">,</span> <span class="n">dbfn</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Set up a :class:`Chipseq` object.</span>


<span class="sd">        :param ip_bam: filename of BAM file for ChIP data</span>
<span class="sd">        :param control_bam: filename of BAM file for control data</span>
<span class="sd">        :param dbfn: filename of gffutils database</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ip</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">genomic_signal</span><span class="p">(</span><span class="n">ip_bam</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;bam&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">control</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">genomic_signal</span><span class="p">(</span><span class="n">control_bam</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s">&#39;bam&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">dbfn</span> <span class="o">=</span> <span class="n">dbfn</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">dbfn</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureDB</span><span class="p">(</span><span class="n">dbfn</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">control_array</span> <span class="o">=</span> <span class="bp">None</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_strip_kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">color</span><span class="o">=</span><span class="s">&#39;.5&#39;</span><span class="p">,</span> <span class="n">markeredgewidth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s">&#39;None&#39;</span><span class="p">,</span>
            <span class="n">picker</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">browser_plotting_kwargs</span> <span class="o">=</span> <span class="p">[</span>
            <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&#39;r&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s">&#39;IP&#39;</span><span class="p">),</span>
            <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s">&#39;control&#39;</span><span class="p">)</span>
        <span class="p">]</span>

</div>
<div class="viewcode-block" id="Chipseq.diff_array"><a class="viewcode-back" href="../../../autodocs/metaseq.integration.chipseq.Chipseq.html#metaseq.integration.chipseq.Chipseq.diff_array">[docs]</a>    <span class="k">def</span> <span class="nf">diff_array</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">features</span><span class="p">,</span> <span class="n">force</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                   <span class="n">array_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(),</span> <span class="n">cache</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Scales the control and IP data to million mapped reads, then subtracts</span>
<span class="sd">        scaled control from scaled IP, applies `func(diffed)` to the diffed</span>
<span class="sd">        array, and finally sets `self.diffed_array` to be the result.</span>

<span class="sd">        Arrays `self.ip` and `self.control` are set as well, and if</span>
<span class="sd">        `force=False`, then previously-created arrays will be used instead of</span>
<span class="sd">        re-calculating new ones.  This is useful if you want to easily try</span>
<span class="sd">        multiple `func` functions without having to re-calculate the data.</span>

<span class="sd">        Another side-effect is that `self.features` is set so that it can be</span>
<span class="sd">        accesed by other methods.</span>

<span class="sd">        :param features: a list of pybedtools.Interval objects</span>
<span class="sd">        :param array_kwargs: extra keyword args passed to genomic_signal.array;</span>
<span class="sd">            typically this will include `bins`, `processes`, and `chunksize`</span>
<span class="sd">            arguments.</span>
<span class="sd">        :param func: a function to apply to the diffed arrays. By default</span>
<span class="sd">            this is :func:`metaseq.plotutils.nice_log`; another option might be</span>
<span class="sd">            `lambda x: x`, or `lambda x: 1e6*x`</span>
<span class="sd">        :param force: Force a re-calculation of the arrays; otherwise uses</span>
<span class="sd">            cached values</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span> <span class="o">=</span> <span class="n">array_kwargs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;processes&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;chunksize&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">array_kwargs</span> <span class="o">=</span> <span class="n">array_kwargs</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="o">**</span><span class="n">array_kwargs</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">/=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="o">.</span><span class="n">mapped_read_count</span><span class="p">()</span> <span class="o">/</span> <span class="mf">1e6</span>

        <span class="k">if</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">control_array</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">)</span> <span class="ow">or</span> <span class="n">force</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">control_array</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">features</span><span class="p">,</span> <span class="o">**</span><span class="n">array_kwargs</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">control_array</span> <span class="o">/=</span> <span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="o">.</span><span class="n">mapped_read_count</span><span class="p">()</span> <span class="o">/</span> <span class="mf">1e6</span>

        <span class="k">if</span> <span class="n">func</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="c">#func = metaseq.plotutils.nice_log</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">control_array</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span> <span class="o">=</span> <span class="n">func</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ip_array</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">control_array</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Chipseq.plot"><a class="viewcode-back" href="../../../autodocs/metaseq.integration.chipseq.Chipseq.html#metaseq.integration.chipseq.Chipseq.plot">[docs]</a>    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">row_order</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">imshow_kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">strip</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot the scaled ChIP-seq data.</span>

<span class="sd">        :param x: X-axis to use (e.g, for TSS +/- 1kb with 100 bins, this would</span>
<span class="sd">            be `np.linspace(-1000, 1000, 100)`)</span>
<span class="sd">        :param row_order: Array-like object containing row order -- typically</span>
<span class="sd">            the result of an `np.argsort` call.</span>
<span class="sd">        :param strip: Include axes along the left side with points that can be</span>
<span class="sd">            clicked to spawn a minibrowser for that feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">nrows</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">row_order</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">row_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">nrows</span><span class="p">)</span>
        <span class="n">extent</span> <span class="o">=</span> <span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nrows</span><span class="p">)</span>
        <span class="n">axes_info</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">plotutils</span><span class="o">.</span><span class="n">matrix_and_line_shell</span><span class="p">(</span><span class="n">strip</span><span class="o">=</span><span class="n">strip</span><span class="p">)</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">matrix_ax</span><span class="p">,</span> <span class="n">line_ax</span><span class="p">,</span> <span class="n">strip_ax</span><span class="p">,</span> <span class="n">cbar_ax</span> <span class="o">=</span> <span class="n">axes_info</span>
        <span class="n">_imshow_kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">aspect</span><span class="o">=</span><span class="s">&#39;auto&#39;</span><span class="p">,</span> <span class="n">extent</span><span class="o">=</span><span class="n">extent</span><span class="p">,</span> <span class="n">interpolation</span><span class="o">=</span><span class="s">&#39;nearest&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">imshow_kwargs</span><span class="p">:</span>
            <span class="n">_imshow_kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">imshow_kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="s">&#39;cmap&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_imshow_kwargs</span><span class="p">:</span>
            <span class="n">_imshow_kwargs</span><span class="p">[</span><span class="s">&#39;cmap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">colormap_adjust</span><span class="o">.</span><span class="n">smart_colormap</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
            <span class="p">)</span>
        <span class="n">mappable</span> <span class="o">=</span> <span class="n">matrix_ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span><span class="p">[</span><span class="n">row_order</span><span class="p">],</span>
            <span class="o">**</span><span class="n">_imshow_kwargs</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">mappable</span><span class="p">,</span> <span class="n">cbar_ax</span><span class="p">)</span>
        <span class="n">line_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">strip_ax</span><span class="p">:</span>
            <span class="n">line</span><span class="p">,</span> <span class="o">=</span> <span class="n">strip_ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">nrows</span><span class="p">,)),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">nrows</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">,</span>
                                  <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">_strip_kwargs</span><span class="p">)</span>
            <span class="n">line</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">features</span>
            <span class="n">line</span><span class="o">.</span><span class="n">ind</span> <span class="o">=</span> <span class="n">row_order</span>

        <span class="n">matrix_ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s">&#39;tight&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">strip_ax</span><span class="p">:</span>
            <span class="n">strip_ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>
            <span class="n">matrix_ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>

        <span class="n">matrix_ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="bp">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">minibrowser</span> <span class="o">=</span> <span class="n">GeneChipseqMiniBrowser</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="p">],</span>
                <span class="n">db</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">,</span>
                <span class="n">plotting_kwargs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">browser_plotting_kwargs</span><span class="p">,</span>
                <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">minibrowser</span> <span class="o">=</span> <span class="n">SignalChipseqMiniBrowser</span><span class="p">(</span>
                <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="p">],</span>
                <span class="n">plotting_kwargs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">browser_plotting_kwargs</span><span class="p">,</span>
                <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">browser_local_coverage_kwargs</span><span class="p">)</span>

        <span class="n">fig</span><span class="o">.</span><span class="n">canvas</span><span class="o">.</span><span class="n">mpl_connect</span><span class="p">(</span><span class="s">&#39;pick_event&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">callback</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span> <span class="o">=</span> <span class="n">fig</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">axes</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s">&#39;matrix_ax&#39;</span><span class="p">:</span> <span class="n">matrix_ax</span><span class="p">,</span>
            <span class="s">&#39;strip_ax&#39;</span><span class="p">:</span> <span class="n">strip_ax</span><span class="p">,</span>
            <span class="s">&#39;line_ax&#39;</span><span class="p">:</span> <span class="n">line_ax</span><span class="p">,</span>
            <span class="s">&#39;cbar_ax&#39;</span><span class="p">:</span> <span class="n">cbar_ax</span>
        <span class="p">}</span>
</div>
<div class="viewcode-block" id="Chipseq.callback"><a class="viewcode-back" href="../../../autodocs/metaseq.integration.chipseq.Chipseq.html#metaseq.integration.chipseq.Chipseq.callback">[docs]</a>    <span class="k">def</span> <span class="nf">callback</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">event</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Callback function to spawn a mini-browser when a feature is clicked.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">artist</span> <span class="o">=</span> <span class="n">event</span><span class="o">.</span><span class="n">artist</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">artist</span><span class="o">.</span><span class="n">ind</span>
        <span class="n">limit</span> <span class="o">=</span> <span class="mi">5</span>
        <span class="n">browser</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">event</span><span class="o">.</span><span class="n">ind</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">limit</span><span class="p">:</span>
            <span class="k">print</span> <span class="s">&quot;more than </span><span class="si">%s</span><span class="s"> genes selected; not spawning browsers&quot;</span> <span class="o">%</span> <span class="n">limit</span>
            <span class="n">browser</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">event</span><span class="o">.</span><span class="n">ind</span><span class="p">:</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="n">artist</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="n">ind</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
            <span class="k">print</span> <span class="n">feature</span><span class="p">,</span>
            <span class="k">if</span> <span class="n">browser</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">minibrowser</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>

</div></div>
<span class="k">def</span> <span class="nf">estimate_shift</span><span class="p">(</span><span class="n">signal</span><span class="p">,</span> <span class="n">genome</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">windowsize</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">thresh</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                   <span class="n">nwindows</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">maxlag</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">array_kwargs</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span>
                   <span class="n">verbose</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Experimental: cross-correlation to estimate the shift width of ChIP-seq</span>
<span class="sd">    data</span>

<span class="sd">    This can be interpreted as the binding site footprint.</span>

<span class="sd">    For ChIP-seq, the plus and minus strand reads tend to be shifted in the 5&#39;</span>
<span class="sd">    direction away from each other.  Various ChIP-seq peak-callers estimate</span>
<span class="sd">    this distance; this function provides a quick, tunable way to do so using</span>
<span class="sd">    cross-correlation.  The resulting shift can then be incorporated into</span>
<span class="sd">    subsequent calls to `array` by adding the shift_width kwarg.</span>


<span class="sd">    :param signal: genomic_signal object</span>
<span class="sd">    :param genome: String assembly for constructing windows</span>
<span class="sd">    :param nwindows: Number of windows to compute cross-correlation on</span>
<span class="sd">    :param windowsize: Size of each window to compute cross-correlation on.</span>
<span class="sd">    :param thresh: Threshold read coverage to run cross-correlation on.  This</span>
<span class="sd">        is likely to be a function of the fragment size provided in</span>
<span class="sd">        `array_kwargs` `windowsize`.  If `thresh` is small, then the cross</span>
<span class="sd">        correlation can be noisy.</span>
<span class="sd">    :param maxlag: Max shift to look for</span>
<span class="sd">    :param array_kwargs: Kwargs passed directly to genomic_signal.array, with</span>
<span class="sd">        the default of `bins=windowsize` for single-bp resolution, and</span>
<span class="sd">        `read_strand` will be overwritten.</span>
<span class="sd">    :param verbose: Be verbose.</span>

<span class="sd">    Returns lags and a `maxlag*2+1` x `nwindows` matrix of cross-correlations.</span>

<span class="sd">    You can then plot the average cross-correlation function with::</span>

<span class="sd">        plt.plot(lags, shift.mean(axis=0))</span>

<span class="sd">    and get the distance to shift with::</span>

<span class="sd">        d = lags[np.argmax(shift.mean(axis=0))]</span>

<span class="sd">    and then plot that with::</span>

<span class="sd">        plt.axvline(d, color=&#39;k&#39;, linestyle=&#39;--&#39;)</span>

<span class="sd">    The number of windows with at least `thresh` coverage is::</span>

<span class="sd">        shift.shape[0]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">thresh</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">thresh</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">if</span> <span class="n">genome</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">genome</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">array_kwargs</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">array_kwargs</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">array_kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s">&#39;read_strand&#39;</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>

    <span class="k">if</span> <span class="s">&#39;bins&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">array_kwargs</span><span class="p">:</span>
        <span class="n">array_kwargs</span><span class="p">[</span><span class="s">&#39;bins&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">windowsize</span>

    <span class="k">def</span> <span class="nf">add_strand</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">strand</span><span class="p">):</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">fields</span><span class="p">[:]</span>
        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
            <span class="n">fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;.&#39;</span><span class="p">)</span>
        <span class="n">fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">strand</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">create_interval_from_list</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span>

    <span class="n">windows</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">()</span>\
        <span class="o">.</span><span class="n">window_maker</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="n">windowsize</span><span class="p">)</span>

    <span class="n">random_subset</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">windows</span><span class="p">[:</span><span class="n">nwindows</span><span class="p">])</span>\
        <span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="n">genome</span><span class="p">)</span><span class="o">.</span><span class="n">saveas</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Getting plus-strand signal for </span><span class="si">%s</span><span class="s"> regions...</span><span class="se">\n</span><span class="s">&quot;</span>
                         <span class="o">%</span> <span class="n">nwindows</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="n">plus</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="n">features</span><span class="o">=</span><span class="n">random_subset</span><span class="p">,</span>
        <span class="n">read_strand</span><span class="o">=</span><span class="s">&quot;+&quot;</span><span class="p">,</span>
        <span class="o">**</span><span class="n">array_kwargs</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Getting minus-strand signal for </span><span class="si">%s</span><span class="s"> regions...</span><span class="se">\n</span><span class="s">&quot;</span>
                         <span class="o">%</span> <span class="n">nwindows</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>

    <span class="n">minus</span> <span class="o">=</span> <span class="n">signal</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="n">features</span><span class="o">=</span><span class="n">random_subset</span><span class="p">,</span>
        <span class="n">read_strand</span><span class="o">=</span><span class="s">&quot;-&quot;</span><span class="p">,</span>
        <span class="o">**</span><span class="n">array_kwargs</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>

    <span class="c"># only do cross-correlation if you have enough reads to do so</span>
    <span class="n">enough</span> <span class="o">=</span> <span class="p">((</span><span class="n">plus</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">windowsize</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">thresh</span><span class="p">)</span> \
        <span class="o">&amp;</span> <span class="p">((</span><span class="n">minus</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">windowsize</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">thresh</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="s">&quot;Running cross-correlation on </span><span class="si">%s</span><span class="s"> regions that passed &quot;</span>
            <span class="s">&quot;threshold</span><span class="se">\n</span><span class="s">&quot;</span> <span class="o">%</span> <span class="nb">sum</span><span class="p">(</span><span class="n">enough</span><span class="p">))</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">sum</span><span class="p">(</span><span class="n">enough</span><span class="p">),</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">maxlag</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">xy</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">izip</span><span class="p">(</span><span class="n">plus</span><span class="p">[</span><span class="n">enough</span><span class="p">],</span> <span class="n">minus</span><span class="p">[</span><span class="n">enough</span><span class="p">])):</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">xy</span>
        <span class="n">results</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">xcorr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">maxlag</span><span class="p">)</span>

    <span class="n">lags</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">maxlag</span><span class="p">,</span> <span class="n">maxlag</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">lags</span><span class="p">,</span> <span class="n">results</span>


<span class="k">def</span> <span class="nf">xcorr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">maxlags</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Streamlined version of matplotlib&#39;s `xcorr`, without the plots.</span>

<span class="sd">    :param x, y: NumPy arrays to cross-correlate</span>
<span class="sd">    :param maxlags: Max number of lags; result will be `2*maxlags+1` in length</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">xlen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="n">ylen</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">xlen</span> <span class="o">==</span> <span class="n">ylen</span>

    <span class="n">c</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">correlate</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="c"># normalize</span>
    <span class="n">c</span> <span class="o">/=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span>

    <span class="n">lags</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="o">-</span><span class="n">maxlags</span><span class="p">,</span> <span class="n">maxlags</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">c</span><span class="p">[</span><span class="n">xlen</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">maxlags</span><span class="p">:</span><span class="n">xlen</span> <span class="o">+</span> <span class="n">maxlags</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">c</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">choices</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;xcorr&#39;</span><span class="p">,</span> <span class="s">&#39;chipseq&#39;</span><span class="p">]</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">examples</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
        <span class="k">print</span> <span class="s">&#39;Choices are: &#39;</span><span class="p">,</span> <span class="n">choices</span>
        <span class="n">examples</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">ex</span> <span class="ow">in</span> <span class="n">examples</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ex</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">choices</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s">&#39;</span><span class="si">%s</span><span class="s"> not in </span><span class="si">%s</span><span class="s">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">ex</span><span class="p">,</span> <span class="n">choices</span><span class="p">))</span>

    <span class="k">if</span> <span class="s">&#39;xcorr&#39;</span> <span class="ow">in</span> <span class="n">examples</span><span class="p">:</span>
        <span class="n">ip</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">genomic_signal</span><span class="p">(</span>
            <span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span>
                <span class="s">&#39;wgEncodeUwTfbsK562CtcfStdAlnRep1.bam&#39;</span><span class="p">),</span> <span class="s">&#39;bam&#39;</span><span class="p">)</span>

        <span class="n">NWINDOWS</span> <span class="o">=</span> <span class="mi">5000</span>
        <span class="n">FRAGMENT_SIZE</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">WINDOWSIZE</span> <span class="o">=</span> <span class="mi">5000</span>
        <span class="n">THRESH</span> <span class="o">=</span> <span class="n">FRAGMENT_SIZE</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">WINDOWSIZE</span><span class="p">)</span> <span class="o">*</span> <span class="mi">10</span>
        <span class="n">lags</span><span class="p">,</span> <span class="n">shift</span> <span class="o">=</span> <span class="n">estimate_shift</span><span class="p">(</span>
            <span class="n">ip</span><span class="p">,</span> <span class="n">nwindows</span><span class="o">=</span><span class="n">NWINDOWS</span><span class="p">,</span> <span class="n">maxlag</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">thresh</span><span class="o">=</span><span class="n">THRESH</span><span class="p">,</span>
            <span class="n">array_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">processes</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="n">fragment_size</span><span class="o">=</span><span class="n">FRAGMENT_SIZE</span><span class="p">),</span>
            <span class="n">verbose</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">lags</span><span class="p">,</span> <span class="n">shift</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span>
            <span class="n">lags</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">shift</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))],</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s">&#39;--&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&#39;k&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s">&#39;chipseq&#39;</span> <span class="ow">in</span> <span class="n">examples</span><span class="p">:</span>
        <span class="c"># Example files...</span>
        <span class="n">dbfn</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span>
            <span class="s">&#39;Homo_sapiens.GRCh37.66.cleaned.gtf.db&#39;</span><span class="p">)</span>
        <span class="n">C</span> <span class="o">=</span> <span class="n">Chipseq</span><span class="p">(</span>
            <span class="n">ip_bam</span><span class="o">=</span><span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span>
                <span class="s">&#39;wgEncodeUwTfbsK562CtcfStdAlnRep1.bam&#39;</span><span class="p">),</span>
            <span class="n">control_bam</span><span class="o">=</span><span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span>
                <span class="s">&#39;wgEncodeUwTfbsK562InputStdAlnRep1.bam&#39;</span><span class="p">),</span>
            <span class="n">dbfn</span><span class="o">=</span><span class="n">dbfn</span><span class="p">)</span>

        <span class="c"># Make some features to use (TSS +/- 1kb)</span>
        <span class="k">def</span> <span class="nf">generator</span><span class="p">():</span>
            <span class="n">G</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureDB</span><span class="p">(</span><span class="n">dbfn</span><span class="p">)</span>
            <span class="n">genes</span> <span class="o">=</span> <span class="n">G</span><span class="o">.</span><span class="n">features_of_type</span><span class="p">(</span><span class="s">&#39;gene&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5000</span><span class="p">):</span>
                <span class="k">yield</span> <span class="n">asinterval</span><span class="p">(</span><span class="n">genes</span><span class="o">.</span><span class="n">next</span><span class="p">())</span>

        <span class="kn">from</span> <span class="nn">pybedtools.featurefuncs</span> <span class="kn">import</span> <span class="n">TSS</span>
        <span class="n">features</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">generator</span><span class="p">())</span>\
            <span class="o">.</span><span class="n">each</span><span class="p">(</span><span class="n">TSS</span><span class="p">,</span> <span class="n">upstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">downstream</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>\
            <span class="o">.</span><span class="n">saveas</span><span class="p">()</span>

        <span class="c"># x-axis for plots</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="o">-</span><span class="mi">500</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>

        <span class="c"># Create the array</span>
        <span class="n">C</span><span class="o">.</span><span class="n">diff_array</span><span class="p">(</span>
            <span class="n">features</span><span class="o">=</span><span class="n">features</span><span class="p">,</span>
            <span class="n">array_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span>
                <span class="n">fragment_size</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">processes</span><span class="o">=</span><span class="mi">6</span><span class="p">))</span>

        <span class="c"># sort genes by TIP zscore</span>
        <span class="n">row_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span>
            <span class="n">metaseq</span><span class="o">.</span><span class="n">plotutils</span><span class="o">.</span><span class="n">tip_zscores</span><span class="p">(</span><span class="n">C</span><span class="o">.</span><span class="n">diffed_array</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

        <span class="c"># Plot &#39;em using a nice red-to-blue colormap</span>
        <span class="kn">from</span> <span class="nn">metaseq.colormap_adjust</span> <span class="kn">import</span> <span class="n">smart_colormap</span>
        <span class="n">cmap</span> <span class="o">=</span> <span class="n">smart_colormap</span><span class="p">(</span><span class="n">C</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">C</span><span class="o">.</span><span class="n">diffed_array</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>

        <span class="n">C</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">row_order</span><span class="o">=</span><span class="n">row_order</span><span class="p">,</span> <span class="n">imshow_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../../index.html">metaseq 0.5.6 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2015, Ryan Dale.
  Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.3.1.
  <br />Theme based on <a href="http://readthedocs.org/">Read The Docs</a>

</div>





  </body>
</html>