<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    

    <title>metaseq.minibrowser &#8212; metaseq 0.5.6 documentation</title>

<meta name="viewport" content="width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=0;"/>


    
    <link rel="stylesheet" href="../../_static/rtd.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '0.5.6',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../../_static/searchtools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">metaseq 0.5.6 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for metaseq.minibrowser</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Module for spawning mini genome browsers using a plugin structure, making it</span>
<span class="sd">possible to build rather complex mini-browsers.  The goal is to point the</span>
<span class="sd">mini-browser to some data, and call its plot() method with a feature.  This</span>
<span class="sd">will spawn a new figure showing the data for that interval.</span>

<span class="sd">MiniBrowser classes are just a general way of mapping data-manipulation or</span>
<span class="sd">data-visualization methods to an Axes on which the data should be displayed.</span>

<span class="sd">To make a new subclass:</span>

<span class="sd">1. Create one or more methods that accept an Axes object and a pybedtools</span>
<span class="sd">   Interval object and return a feature.  The simplest do-nothing method would</span>
<span class="sd">   be::</span>

<span class="sd">        def my_panel(self, ax, feature)</span>
<span class="sd">            return feature</span>

<span class="sd">   A more useful method might be one that plots genomic signal over the</span>
<span class="sd">   region::</span>

<span class="sd">        def my_panel(self, ax, feature):</span>
<span class="sd">            # for simplicity, assume just use the first genomic_signal</span>
<span class="sd">            gs = self.genomic_signal_objs[0]</span>
<span class="sd">            x, y = gs.local_coverage(feature, bins=100)</span>
<span class="sd">            ax.plot(x, y, **kwargs)</span>
<span class="sd">            ax.axis(&#39;tight&#39;)</span>
<span class="sd">            return feature</span>

<span class="sd">2. Then, override the `panels()` method.  This method:</span>

<span class="sd">    * Creates Axes as needed; assumes that self.make_fig() has already been</span>
<span class="sd">      called so that self.fig is available.</span>
<span class="sd">    * Returns a list of (ax, method) tuples.  This list maps created Axes to</span>
<span class="sd">      methods that should operate on them (like `my_panel` method above).</span>

<span class="sd">    For example::</span>

<span class="sd">        def panels(self):</span>
<span class="sd">            ax = self.fig.add_subplot(111)</span>
<span class="sd">            return [(ax, self.my_panel)]</span>


<span class="sd">A figure is spawned by calling the `plot` method on a pybedtools genomic</span>
<span class="sd">interval, e.g.::</span>

<span class="sd">    s = SignalMiniBrowser(ip, control])</span>
<span class="sd">    s.plot(feature)</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="k">import</span> <span class="n">copy</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="k">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">from</span> <span class="nn">pybedtools.contrib.plotting</span> <span class="k">import</span> <span class="n">Track</span>
<span class="kn">import</span> <span class="nn">pybedtools</span>
<span class="kn">import</span> <span class="nn">gffutils</span>
<span class="kn">from</span> <span class="nn">gffutils.helpers</span> <span class="k">import</span> <span class="n">asinterval</span>
<span class="kn">from</span> <span class="nn">mpl_toolkits.axes_grid1</span> <span class="k">import</span> <span class="n">make_axes_locatable</span>
<span class="kn">import</span> <span class="nn">matplotlib.gridspec</span> <span class="k">as</span> <span class="nn">gridspec</span>
<span class="kn">from</span> <span class="nn">matplotlib.ticker</span> <span class="k">import</span> <span class="n">FormatStrFormatter</span>
<span class="kn">import</span> <span class="nn">_genomic_signal</span>


<div class="viewcode-block" id="BaseMiniBrowser"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser">[docs]</a><span class="k">class</span> <span class="nc">BaseMiniBrowser</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Base class for plotting a genomic region.</span>

<span class="sd">    This class is designed to be sub-classed, so it really just a shell with</span>
<span class="sd">    some example methods filled in.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_all_figures</span> <span class="o">=</span> <span class="p">[]</span>

<div class="viewcode-block" id="BaseMiniBrowser.__init__"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genomic_signal_objs</span> <span class="o">=</span> <span class="n">genomic_signal_objs</span></div>

<div class="viewcode-block" id="BaseMiniBrowser.close_all"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.close_all">[docs]</a>    <span class="k">def</span> <span class="nf">close_all</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Close all figures spawned by this class.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_all_figures</span><span class="p">:</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">f</span><span class="p">)</span></div>

<div class="viewcode-block" id="BaseMiniBrowser.plot"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.plot">[docs]</a>    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Spawns a new figure showing data for `feature`.</span>

<span class="sd">        :param feature: A `pybedtools.Interval` object</span>

<span class="sd">        Using the pybedtools.Interval `feature`, creates figure specified in</span>
<span class="sd">        :meth:`BaseMiniBrowser.make_fig` and plots data on panels according to</span>
<span class="sd">        `self.panels()`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">Feature</span><span class="p">):</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="n">asinterval</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_fig</span><span class="p">()</span>
        <span class="n">axes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">ax</span><span class="p">,</span> <span class="n">method</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">panels</span><span class="p">():</span>
            <span class="n">feature</span> <span class="o">=</span> <span class="n">method</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span>
            <span class="n">axes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">axes</span></div>

<div class="viewcode-block" id="BaseMiniBrowser.make_fig"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.make_fig">[docs]</a>    <span class="k">def</span> <span class="nf">make_fig</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Figure constructor, called before `self.plot()`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_all_figures</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="p">)</span></div>

<div class="viewcode-block" id="BaseMiniBrowser.panels"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.panels">[docs]</a>    <span class="k">def</span> <span class="nf">panels</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Method to be overriden by subclasses.</span>

<span class="sd">        This method should create Axes objects on self.fig and register which</span>
<span class="sd">        methods operate on which axes.</span>

<span class="sd">        Each method&#39;s signature should be func(self, ax, feature), and it will</span>
<span class="sd">        take a pybedtools.Interval `feature` and plot it or something derived</span>
<span class="sd">        from it onto `ax`. It should return a pybedtools.Interval object,</span>
<span class="sd">        usually just the same one that was passed in.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[(</span><span class="n">ax</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">example_panel</span><span class="p">)]</span></div>

<div class="viewcode-block" id="BaseMiniBrowser.example_panel"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.BaseMiniBrowser.html#metaseq.minibrowser.BaseMiniBrowser.example_panel">[docs]</a>    <span class="k">def</span> <span class="nf">example_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A example panel that just prints the text of the feature.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">txt</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%s</span><span class="s1">-</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">txt</span><span class="p">,</span> <span class="n">transform</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">transAxes</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feature</span></div></div>


<span class="k">class</span> <span class="nc">ChIPSeqMiniBrowser</span><span class="p">(</span><span class="n">BaseMiniBrowser</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ip</span><span class="p">,</span> <span class="n">control</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">peaks</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">stranded</span><span class="o">=</span><span class="kc">False</span><span class="p">),</span> <span class="n">ip_style</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">control_style</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">peaks_style</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">max_bins</span><span class="o">=</span><span class="mi">500</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Mini-browser to show IP and control signal, along with optional peaks</span>
<span class="sd">        and gene models.  See module-level documentation for details.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ip, control : genomic_signal object</span>
<span class="sd">            These can be any kind of genomic signal object.  If BamSignal, then</span>
<span class="sd">            the signal will be scaled by million mapped reads.</span>

<span class="sd">        db : filename or gffutils.FeatureDB</span>
<span class="sd">            Optional database of annotations.  If provided, gene models will be</span>
<span class="sd">            plotted on an additional axes.</span>

<span class="sd">        peaks : filename or pybedtools.BedTool object</span>
<span class="sd">            Optional file of called peaks.  If provided, will be plotted on an</span>
<span class="sd">            additional axes.</span>

<span class="sd">        local_coverage_kwargs : dict</span>
<span class="sd">            Kwargs to pass on to the local_coverage method of `ip` and</span>
<span class="sd">            `control`.  It is recommended that at least `stranded=False` is</span>
<span class="sd">            included, since the mini-browser always displays the plus strand.</span>

<span class="sd">        ip_style, control_style : dict</span>
<span class="sd">            Kwargs to pass to Axes.fill_between (e.g. colors, alpha, line</span>
<span class="sd">            styles)</span>

<span class="sd">        peaks_style : dict</span>
<span class="sd">            Kwargs to pass to PolyCollection (e.g., face_color, edgewidth)</span>

<span class="sd">        max_bins : int</span>
<span class="sd">            Maximum number of bins to use when getting genomic_signal.  This</span>
<span class="sd">            can be overridden by providing the `bins` kwarg in</span>
<span class="sd">            `local_coverage_kwargs`.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        After creation, settings can be changed in between calls to the</span>
<span class="sd">        `plot()` method.</span>

<span class="sd">        For example, the self.settings dictionary, which</span>
<span class="sd">        contains arguments passed to the gffutils.contrib.plotting.Gene class,</span>
<span class="sd">        can be modified to specify which kinds of features should be plotted.</span>
<span class="sd">        The `gs` attribute contains the matplotlib.gridspec.GridSpec object,</span>
<span class="sd">        which can be used to modify the relative sizes of the subplots.</span>


<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">ChIPSeqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">([</span><span class="n">ip</span><span class="p">,</span> <span class="n">control</span><span class="p">])</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">local_coverage_kwargs</span> <span class="o">=</span> <span class="n">local_coverage_kwargs</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ip</span> <span class="o">=</span> <span class="n">ip</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">control</span> <span class="o">=</span> <span class="n">control</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">db</span><span class="p">,</span> <span class="n">basestring</span><span class="p">):</span>
            <span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureDB</span><span class="p">(</span><span class="n">db</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">db</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ip_style</span> <span class="o">=</span> <span class="n">ip_style</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peaks_style</span> <span class="o">=</span> <span class="n">peaks_style</span> <span class="ow">or</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">control_style</span> <span class="o">=</span> <span class="n">control_style</span> <span class="ow">or</span> <span class="p">{}</span>

        <span class="k">if</span> <span class="n">peaks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">peaks</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">peaks</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span> <span class="o">=</span> <span class="n">peaks</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_bins</span> <span class="o">=</span> <span class="n">max_bins</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">settings</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;transcripts&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="s1">&#39;cds&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CDS&#39;</span><span class="p">],</span>
            <span class="s1">&#39;utrs&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;exon&#39;</span><span class="p">],</span>
            <span class="s1">&#39;color&#39;</span><span class="p">:</span> <span class="s1">&#39;0.5&#39;</span><span class="p">,</span>
            <span class="s1">&#39;featuretype&#39;</span><span class="p">:</span> <span class="s1">&#39;gene&#39;</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="k">def</span> <span class="nf">panels</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
        <span class="n">method_dispatch</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s1">&#39;ip&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">ip_panel</span><span class="p">,</span>
            <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">control_panel</span><span class="p">,</span>
            <span class="s1">&#39;peaks&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks_panel</span><span class="p">,</span>
            <span class="s1">&#39;genes&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">genes_panel</span><span class="p">}</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span><span class="p">:</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">15</span><span class="p">,</span> <span class="o">.</span><span class="mi">5</span><span class="p">])</span>
            <span class="n">ip_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">control_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">peaks_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">gene_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">axes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ip&#39;</span><span class="p">:</span> <span class="n">ip_ax</span><span class="p">,</span>
                    <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="n">control_ax</span><span class="p">,</span>
                    <span class="s1">&#39;peaks&#39;</span><span class="p">:</span> <span class="n">peaks_ax</span><span class="p">,</span>
                    <span class="s1">&#39;genes&#39;</span><span class="p">:</span> <span class="n">gene_ax</span><span class="p">,</span>
                    <span class="p">}</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">5</span><span class="p">])</span>
            <span class="n">ip_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">control_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">gene_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">axes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ip&#39;</span><span class="p">:</span> <span class="n">ip_ax</span><span class="p">,</span>
                    <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="n">control_ax</span><span class="p">,</span>
                    <span class="s1">&#39;genes&#39;</span><span class="p">:</span> <span class="n">gene_ax</span><span class="p">,</span>
                    <span class="p">}</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span><span class="p">:</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="o">.</span><span class="mi">3</span><span class="p">])</span>
            <span class="n">ip_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">control_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">peaks_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">axes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ip&#39;</span><span class="p">:</span> <span class="n">ip_ax</span><span class="p">,</span>
                    <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="n">control_ax</span><span class="p">,</span>
                    <span class="s1">&#39;peaks&#39;</span><span class="p">:</span> <span class="n">peaks_ax</span><span class="p">,</span>
                    <span class="p">}</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">gs</span> <span class="o">=</span> <span class="n">gridspec</span><span class="o">.</span><span class="n">GridSpec</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">height_ratios</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
            <span class="n">ip_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">control_ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="n">gs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">ip_ax</span><span class="p">)</span>
            <span class="n">axes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ip&#39;</span><span class="p">:</span> <span class="n">ip_ax</span><span class="p">,</span>
                    <span class="s1">&#39;control&#39;</span><span class="p">:</span> <span class="n">control_ax</span><span class="p">,</span>
                    <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">axes</span> <span class="o">=</span> <span class="n">axes</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gridspec</span> <span class="o">=</span> <span class="n">gs</span>

        <span class="c1"># Order shouldn&#39;t matter, but just in case...</span>
        <span class="n">mapping</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">kind</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ip&#39;</span><span class="p">,</span> <span class="s1">&#39;control&#39;</span><span class="p">,</span> <span class="s1">&#39;peaks&#39;</span><span class="p">,</span> <span class="s1">&#39;genes&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">kind</span> <span class="ow">in</span> <span class="n">axes</span><span class="p">:</span>
                <span class="n">mapping</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">axes</span><span class="p">[</span><span class="n">kind</span><span class="p">],</span> <span class="n">method_dispatch</span><span class="p">[</span><span class="n">kind</span><span class="p">]))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_first_ax</span> <span class="o">=</span> <span class="n">mapping</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_last_ax</span> <span class="o">=</span> <span class="n">mapping</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">mapping</span>

    <span class="k">def</span> <span class="nf">_bins</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">feature</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_bins</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_zoomed_feature</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="n">extra</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">settings</span><span class="p">[</span><span class="s1">&#39;zoom&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">feature</span> <span class="o">=</span> <span class="n">copy</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">feature</span><span class="o">.</span><span class="n">start</span> <span class="o">-=</span> <span class="n">extra</span>
        <span class="n">feature</span><span class="o">.</span><span class="n">stop</span> <span class="o">+=</span> <span class="n">extra</span>
        <span class="k">return</span> <span class="n">feature</span>

    <span class="k">def</span> <span class="nf">ip_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bins</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="o">.</span><span class="n">local_coverage</span><span class="p">(</span>
            <span class="n">feature</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="p">,</span> <span class="n">_genomic_signal</span><span class="o">.</span><span class="n">BamSignal</span><span class="p">):</span>
            <span class="n">y</span> <span class="o">/=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ip</span><span class="o">.</span><span class="n">mapped_read_count</span><span class="p">()</span> <span class="o">/</span> <span class="mi">1</span><span class="n">e6</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">y2</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">ip_style</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;bottom&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feature</span>

    <span class="k">def</span> <span class="nf">control_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="n">bins</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bins</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="o">.</span><span class="n">local_coverage</span><span class="p">(</span>
            <span class="n">feature</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="p">,</span> <span class="n">_genomic_signal</span><span class="o">.</span><span class="n">BamSignal</span><span class="p">):</span>
            <span class="n">y</span> <span class="o">/=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">control</span><span class="o">.</span><span class="n">mapped_read_count</span><span class="p">()</span> <span class="o">/</span> <span class="mi">1</span><span class="n">e6</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">y2</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">control_style</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;bottom&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feature</span>

    <span class="k">def</span> <span class="nf">peaks_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="n">hits</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">peaks</span><span class="o">.</span><span class="n">intersect</span><span class="p">([</span><span class="n">feature</span><span class="p">],</span> <span class="n">u</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">track</span> <span class="o">=</span> <span class="n">Track</span><span class="p">(</span><span class="n">hits</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">peaks_style</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">track</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">(</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklines</span><span class="p">()</span> <span class="o">+</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklines</span><span class="p">()</span> <span class="o">+</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">()</span>
            <span class="o">+</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">()</span>
        <span class="p">):</span>
            <span class="n">x</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Peaks&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span>
                      <span class="n">verticalalignment</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_frame_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">feature</span>

    <span class="k">def</span> <span class="nf">plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_feature</span> <span class="o">=</span> <span class="n">feature</span>
        <span class="n">axes</span> <span class="o">=</span> <span class="nb">super</span><span class="p">(</span><span class="n">ChIPSeqMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">ax</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">axes</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">_last_ax</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">txt</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_xticklabels</span><span class="p">():</span>
                    <span class="n">txt</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">wspace</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_first_ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="n">xmin</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_last_ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_last_ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_major_formatter</span><span class="p">(</span><span class="n">FormatStrFormatter</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">axes</span>

    <span class="k">def</span> <span class="nf">coords</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_feature</span><span class="o">.</span><span class="n">chrom</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_first_ax</span><span class="o">.</span><span class="n">axis</span><span class="p">()</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">:</span><span class="si">%s</span><span class="s1">-</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">stop</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">genes_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots gene models on an Axes.</span>

<span class="sd">        Queries the database</span>

<span class="sd">        :param ax: matplotlib.Axes object</span>
<span class="sd">        :param feature: pybedtools.Interval</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">gffutils.contrib.plotting</span> <span class="k">import</span> <span class="n">Gene</span>
        <span class="n">extent</span> <span class="o">=</span> <span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">]</span>
        <span class="n">nearby_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">region</span><span class="p">(</span>
            <span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">),</span> <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>
        <span class="n">ybase</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ngenes</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">nearby_gene</span> <span class="ow">in</span> <span class="n">nearby_genes</span><span class="p">:</span>
            <span class="c1"># TODO: there should be a better way of identifying which gene is</span>
            <span class="c1"># the same as the feature requested.  Might need to expose an &quot;ID&quot;</span>
            <span class="c1"># kwarg.</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">nearby_gene</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">feature</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]:</span>
                    <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;0.2&#39;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;0.5&#39;</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;0.5&#39;</span>
            <span class="n">ngenes</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">extent</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">nearby_gene</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">nearby_gene</span><span class="o">.</span><span class="n">stop</span><span class="p">])</span>
            <span class="n">gene_collection</span> <span class="o">=</span> <span class="n">Gene</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">,</span>
                <span class="n">nearby_gene</span><span class="p">,</span>
                <span class="n">transcripts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">cds</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CDS&#39;</span><span class="p">],</span>
                <span class="n">utrs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;exon&#39;</span><span class="p">],</span>
                <span class="n">ybase</span><span class="o">=</span><span class="n">ybase</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">)</span>
            <span class="n">gene_collection</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">nearby_gene</span><span class="o">.</span><span class="n">id</span>
            <span class="n">gene_collection</span><span class="o">.</span><span class="n">add_to_ax</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
            <span class="n">ybase</span> <span class="o">+=</span> <span class="n">gene_collection</span><span class="o">.</span><span class="n">max_y</span>

        <span class="n">xmin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">extent</span><span class="p">)</span>
        <span class="n">xmax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">extent</span><span class="p">)</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">ngenes</span>

        <span class="c1"># 1% padding seems to work well</span>
        <span class="n">padding</span> <span class="o">=</span> <span class="p">(</span><span class="n">xmax</span> <span class="o">-</span> <span class="n">xmin</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.01</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

        <span class="c1"># add lines indicating extent of current feature</span>
        <span class="c1"># vline_kwargs = dict(color=&#39;k&#39;, linestyle=&#39;--&#39;)</span>
        <span class="c1"># ax.axvline(feature.start, **vline_kwargs)</span>
        <span class="c1"># ax.axvline(feature.stop, **vline_kwargs)</span>

        <span class="c1"># Make a new feature to represent the region plus surrounding genes</span>
        <span class="n">interval</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">create_interval_from_list</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">fields</span><span class="p">)</span>
        <span class="n">interval</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span>
        <span class="k">for</span> <span class="n">txt</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklabels</span><span class="p">():</span>
            <span class="n">txt</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">tick</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_yticklines</span><span class="p">():</span>
            <span class="n">tick</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;right&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;left&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="p">[</span><span class="s1">&#39;top&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks_position</span><span class="p">(</span><span class="s1">&#39;bottom&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Genes&#39;</span><span class="p">,</span> <span class="n">rotation</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">horizontalalignment</span><span class="o">=</span><span class="s1">&#39;right&#39;</span><span class="p">,</span>
                      <span class="n">verticalalignment</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">interval</span>


<div class="viewcode-block" id="SignalMiniBrowser"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.SignalMiniBrowser.html#metaseq.minibrowser.SignalMiniBrowser">[docs]</a><span class="k">class</span> <span class="nc">SignalMiniBrowser</span><span class="p">(</span><span class="n">BaseMiniBrowser</span><span class="p">):</span>
<div class="viewcode-block" id="SignalMiniBrowser.__init__"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.SignalMiniBrowser.html#metaseq.minibrowser.SignalMiniBrowser.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">plotting_kwargs</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Base class for plotting genomic signal.</span>

<span class="sd">        Plots genomic signal over a particular area of the genome.  Designed to</span>
<span class="sd">        be extended by subclasses, but can stand alone if all you want is</span>
<span class="sd">        a simple one-panel browser.</span>

<span class="sd">        :param genomic_signal_objs: list of genomic signal objects (e.g.,</span>
<span class="sd">            :class:`metaseq.genomic_signal.BamSignal` instances).</span>
<span class="sd">        :param local_coverage_kwargs: a dictionary of kwargs to send to each</span>
<span class="sd">            genomic signals&#39; `local_coverage()` method, e.g.::</span>

<span class="sd">                local_coverage_kwargs = dict(fragment_size=200).</span>
<span class="sd">        :param plotting_kwargs:  a list of dictionaries, one for each genomic</span>
<span class="sd">            signals object, e.g,::</span>

<span class="sd">                plotting_kwargs = [dict(color=&#39;r&#39;, label=&#39;IP&#39;, dict(color=&#39;k&#39;,</span>
<span class="sd">                    label=&#39;input&#39;)]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">SignalMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="n">genomic_signal_objs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plotting_kwargs</span> <span class="o">=</span> \
            <span class="n">plotting_kwargs</span> <span class="ow">or</span> <span class="p">[{}</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genomic_signal_objs</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">local_coverage_kwargs</span> <span class="o">=</span> <span class="n">local_coverage_kwargs</span> <span class="ow">or</span> <span class="p">{}</span></div>

<div class="viewcode-block" id="SignalMiniBrowser.panels"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.SignalMiniBrowser.html#metaseq.minibrowser.SignalMiniBrowser.panels">[docs]</a>    <span class="k">def</span> <span class="nf">panels</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add a single panel to the figure</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ax</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[(</span><span class="n">ax</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">signal_panel</span><span class="p">)]</span></div>

<div class="viewcode-block" id="SignalMiniBrowser.signal_panel"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.SignalMiniBrowser.html#metaseq.minibrowser.SignalMiniBrowser.signal_panel">[docs]</a>    <span class="k">def</span> <span class="nf">signal_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots each genomic signal as a line using the corresponding</span>
<span class="sd">        plotting_kwargs</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">gs</span><span class="p">,</span> <span class="n">kwargs</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">plotting_kwargs</span><span class="p">):</span>
            <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">gs</span><span class="o">.</span><span class="n">local_coverage</span><span class="p">(</span><span class="n">feature</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">feature</span></div></div>


<div class="viewcode-block" id="GeneModelMiniBrowser"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.GeneModelMiniBrowser.html#metaseq.minibrowser.GeneModelMiniBrowser">[docs]</a><span class="k">class</span> <span class="nc">GeneModelMiniBrowser</span><span class="p">(</span><span class="n">SignalMiniBrowser</span><span class="p">):</span>
<div class="viewcode-block" id="GeneModelMiniBrowser.__init__"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.GeneModelMiniBrowser.html#metaseq.minibrowser.GeneModelMiniBrowser.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="n">db</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Mini-browser to show a signal panel on top and gene models on the</span>
<span class="sd">        bottom.</span>

<span class="sd">        :param genomic_signal_objs: a list of genomic_signals objects</span>
<span class="sd">        :param db: a `gffutils.FeatureDB`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">GeneModelMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span>
            <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db</span> <span class="o">=</span> <span class="n">db</span></div>

<div class="viewcode-block" id="GeneModelMiniBrowser.panels"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.GeneModelMiniBrowser.html#metaseq.minibrowser.GeneModelMiniBrowser.panels">[docs]</a>    <span class="k">def</span> <span class="nf">panels</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add 2 panels to the figure, top for signal and bottom for gene models</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ax1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">211</span><span class="p">)</span>
        <span class="n">ax2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">212</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">ax2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">gene_panel</span><span class="p">),</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">signal_panel</span><span class="p">)</span></div>

<div class="viewcode-block" id="GeneModelMiniBrowser.gene_panel"><a class="viewcode-back" href="../../autodocs/metaseq.minibrowser.GeneModelMiniBrowser.html#metaseq.minibrowser.GeneModelMiniBrowser.gene_panel">[docs]</a>    <span class="k">def</span> <span class="nf">gene_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plots gene models on an Axes.</span>

<span class="sd">        Queries the database</span>

<span class="sd">        :param ax: matplotlib.Axes object</span>
<span class="sd">        :param feature: pybedtools.Interval</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">gffutils.contrib.plotting</span> <span class="k">import</span> <span class="n">Gene</span>
        <span class="n">extent</span> <span class="o">=</span> <span class="p">[</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">]</span>
        <span class="n">nearby_genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="o">.</span><span class="n">region</span><span class="p">(</span>
            <span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">),</span> <span class="n">featuretype</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span><span class="p">)</span>
        <span class="n">ybase</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">ngenes</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">nearby_gene</span> <span class="ow">in</span> <span class="n">nearby_genes</span><span class="p">:</span>
            <span class="n">ngenes</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">extent</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">nearby_gene</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="n">nearby_gene</span><span class="o">.</span><span class="n">stop</span><span class="p">])</span>
            <span class="n">gene_collection</span> <span class="o">=</span> <span class="n">Gene</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db</span><span class="p">,</span>
                <span class="n">nearby_gene</span><span class="p">,</span>
                <span class="n">transcripts</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mRNA&#39;</span><span class="p">],</span>
                <span class="n">cds</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CDS&#39;</span><span class="p">],</span>
                <span class="n">utrs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;exon&#39;</span><span class="p">],</span>
                <span class="n">ybase</span><span class="o">=</span><span class="n">ybase</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="s2">&quot;0.5&quot;</span><span class="p">,</span> <span class="n">picker</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
            <span class="n">gene_collection</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">nearby_gene</span><span class="o">.</span><span class="n">id</span>
            <span class="n">gene_collection</span><span class="o">.</span><span class="n">add_to_ax</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
            <span class="n">ybase</span> <span class="o">+=</span> <span class="n">gene_collection</span><span class="o">.</span><span class="n">max_y</span>

        <span class="n">xmin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">extent</span><span class="p">)</span>
        <span class="n">xmax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">extent</span><span class="p">)</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">ngenes</span>

        <span class="c1"># 1% padding seems to work well</span>
        <span class="n">padding</span> <span class="o">=</span> <span class="p">(</span><span class="n">xmax</span> <span class="o">-</span> <span class="n">xmin</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.01</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;tight&#39;</span><span class="p">)</span>

        <span class="c1"># add lines indicating extent of current feature</span>
        <span class="n">vline_kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">start</span><span class="p">,</span> <span class="o">**</span><span class="n">vline_kwargs</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">stop</span><span class="p">,</span> <span class="o">**</span><span class="n">vline_kwargs</span><span class="p">)</span>

        <span class="c1"># Make a new feature to represent the region plus surrounding genes</span>
        <span class="n">interval</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">create_interval_from_list</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">fields</span><span class="p">)</span>
        <span class="n">interval</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="n">xmin</span> <span class="o">-</span> <span class="n">padding</span>
        <span class="n">interval</span><span class="o">.</span><span class="n">stop</span> <span class="o">=</span> <span class="n">xmax</span> <span class="o">+</span> <span class="n">padding</span>
        <span class="n">interval</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span>
        <span class="k">return</span> <span class="n">interval</span></div></div>


<span class="k">class</span> <span class="nc">PeakMiniBrowser</span><span class="p">(</span><span class="n">SignalMiniBrowser</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="n">bed</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Signal on the top panel, peaks in the bottom panel. `bed` is a filename</span>
<span class="sd">        or BedTool object (or list of these things) that will be used to make</span>
<span class="sd">        a pybedtools.contrib.plotting.Track.</span>

<span class="sd">        If genomic_signal_objs is None, then only show the peaks axes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">PeakMiniBrowser</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="n">genomic_signal_objs</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bed</span> <span class="o">=</span> <span class="n">bed</span>

    <span class="k">def</span> <span class="nf">panels</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">ax1</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">211</span><span class="p">)</span>
        <span class="n">ax2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">212</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">signal_panel</span><span class="p">),</span> <span class="p">(</span><span class="n">ax2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">peak_panel</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">peak_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">feature</span><span class="p">):</span>
        <span class="n">bedtool</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bed</span><span class="p">)</span>
        <span class="n">features</span> <span class="o">=</span> <span class="n">bedtool</span><span class="o">.</span><span class="n">intersect</span><span class="p">([</span><span class="n">feature</span><span class="p">],</span> <span class="n">u</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">track</span> <span class="o">=</span> <span class="n">Track</span><span class="p">(</span><span class="n">features</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">add_collection</span><span class="p">(</span><span class="n">track</span><span class="p">)</span>
        <span class="c1"># ax.axis(&#39;tight&#39;)</span>
        <span class="k">return</span> <span class="n">feature</span>


<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">metaseq</span>
    <span class="kn">import</span> <span class="nn">gffutils</span>
    <span class="kn">import</span> <span class="nn">pybedtools</span>

    <span class="n">G</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">FeatureDB</span><span class="p">(</span>
        <span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s1">&#39;Homo_sapiens.GRCh37.66.cleaned.gtf.db&#39;</span><span class="p">))</span>

    <span class="n">ip</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">genomic_signal</span><span class="p">(</span>
        <span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s1">&#39;wgEncodeUwTfbsK562CtcfStdAlnRep1.bam&#39;</span><span class="p">),</span>
        <span class="s1">&#39;bam&#39;</span><span class="p">)</span>
    <span class="n">inp</span> <span class="o">=</span> <span class="n">metaseq</span><span class="o">.</span><span class="n">genomic_signal</span><span class="p">(</span>
        <span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s1">&#39;wgEncodeUwTfbsK562InputStdAlnRep1.bam&#39;</span><span class="p">),</span>
        <span class="s1">&#39;bam&#39;</span><span class="p">)</span>
    <span class="n">peaks</span> <span class="o">=</span> <span class="n">pybedtools</span><span class="o">.</span><span class="n">BedTool</span><span class="p">(</span><span class="n">metaseq</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span>
        <span class="s1">&#39;wgEncodeUwTfbsK562CtcfStdPkRep1.narrowPeak.gz&#39;</span><span class="p">))</span>

    <span class="n">plotting_kwargs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;IP&#39;</span><span class="p">),</span>
        <span class="nb">dict</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;input&#39;</span><span class="p">)]</span>

    <span class="n">local_coverage_kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">fragment_size</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>

    <span class="n">b</span> <span class="o">=</span> <span class="n">SignalMiniBrowser</span><span class="p">(</span>
        <span class="p">[</span><span class="n">ip</span><span class="p">,</span> <span class="n">inp</span><span class="p">],</span>
        <span class="n">plotting_kwargs</span><span class="o">=</span><span class="n">plotting_kwargs</span><span class="p">,</span>
        <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>

    <span class="n">g</span> <span class="o">=</span> <span class="n">GeneModelMiniBrowser</span><span class="p">(</span>
        <span class="p">[</span><span class="n">ip</span><span class="p">,</span> <span class="n">inp</span><span class="p">],</span>
        <span class="n">G</span><span class="p">,</span>
        <span class="n">plotting_kwargs</span><span class="o">=</span><span class="n">plotting_kwargs</span><span class="p">,</span>
        <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>

    <span class="n">p</span> <span class="o">=</span> <span class="n">PeakMiniBrowser</span><span class="p">(</span>
        <span class="p">[</span><span class="n">ip</span><span class="p">,</span> <span class="n">inp</span><span class="p">],</span>
        <span class="n">peaks</span><span class="p">,</span>
        <span class="n">plotting_kwargs</span><span class="o">=</span><span class="n">plotting_kwargs</span><span class="p">,</span>
        <span class="n">local_coverage_kwargs</span><span class="o">=</span><span class="n">local_coverage_kwargs</span><span class="p">)</span>

    <span class="n">feature</span> <span class="o">=</span> <span class="n">peaks</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

    <span class="n">b</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
    <span class="n">g</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
    <span class="n">p</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">feature</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../../index.html">metaseq 0.5.6 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2015, Ryan Dale.
  Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.5.1.
  <br />Theme based on <a href="http://readthedocs.org/">Read The Docs</a>

</div>





  </body>
</html>